
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <meta name="HandheldFriendly" content="True" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta name="robots" content="" />

  <link href="https://fonts.googleapis.com/css2?family=Source+Code+Pro:ital,wght@0,400;0,700;1,400&family=Source+Sans+Pro:ital,wght@0,300;0,400;0,700;1,400&display=swap" rel="stylesheet">

    <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/stylesheet/style.min.css">


    <link id="pygments-light-theme" rel="stylesheet" type="text/css"
          href="https://antoniocampos13.github.io/theme/pygments/vs.min.css">


  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/fontawesome.css">
  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/brands.css">
  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/solid.css">


    <link href="https://antoniocampos13.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="Antonio's Portfolio Atom">

    <link href="https://antoniocampos13.github.io/feeds/all.rss.xml" type="application/rss+xml" rel="alternate" title="Antonio's Portfolio RSS">



    <!-- Chrome, Firefox OS and Opera -->
    <meta name="theme-color" content="#333333">
    <!-- Windows Phone -->
    <meta name="msapplication-navbutton-color" content="#333333">
    <!-- iOS Safari -->
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    <!-- Microsoft EDGE -->
    <meta name="msapplication-TileColor" content="#333333">

<meta name="author" content="Antonio Victor Campos Coelho" />
<meta name="description" content="Introduction In my previous post, I showed how to configure an Ubuntu system to install Bioinformatics programs. Now, using the environment I created, I will demonstrate a bash script, FastQ_to_Annotation.sh that takes next generation sequencing (NGS) raw reads from human whole genome sequencing as input and produces variant annotation …" />
<meta name="keywords" content="Bioinformatics, genomic variation, entrez-direct, EDirect">


<meta property="og:site_name" content="Antonio's Portfolio"/>
<meta property="og:title" content="FASTQ to Annotation (Part 1)"/>
<meta property="og:description" content="Introduction In my previous post, I showed how to configure an Ubuntu system to install Bioinformatics programs. Now, using the environment I created, I will demonstrate a bash script, FastQ_to_Annotation.sh that takes next generation sequencing (NGS) raw reads from human whole genome sequencing as input and produces variant annotation …"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:url" content="https://antoniocampos13.github.io/fastq-to-annotation-part-1.html"/>
<meta property="og:type" content="article"/>
<meta property="article:published_time" content="2020-10-01 18:00:00-03:00"/>
<meta property="article:modified_time" content=""/>
<meta property="article:author" content="https://antoniocampos13.github.io/author/antonio-victor-campos-coelho.html">
<meta property="article:section" content="Unix"/>
<meta property="article:tag" content="Bioinformatics"/>
<meta property="article:tag" content="genomic variation"/>
<meta property="article:tag" content="entrez-direct"/>
<meta property="article:tag" content="EDirect"/>
<meta property="og:image" content="https://avatars.githubusercontent.com/antoniocampos13">

  <title>Antonio's Portfolio &ndash; FASTQ to Annotation (Part 1)</title>

</head>
<body class="light-theme">
  <aside>
    <div>
      <a href="https://antoniocampos13.github.io">
        <img src="https://avatars.githubusercontent.com/antoniocampos13" alt="Antonio's Portfolio" title="Antonio's Portfolio">
      </a>

      <h1>
        <a href="https://antoniocampos13.github.io">Antonio's Portfolio</a>
      </h1>

<p>PhD in Genetics</p>

      <nav>
        <ul class="list">


              <li>
                <a target="_self"
                   href="https://antoniocampos13.github.io/pages/about.html#about">
                  About
                </a>
              </li>
              <li>
                <a target="_self"
                   href="https://antoniocampos13.github.io/pages/contact.html#contact">
                  Contact
                </a>
              </li>

            <li>
              <a target="_self" href="http://lattes.cnpq.br/2986394950644755" >Brazilian Lattes CV</a>
            </li>
            <li>
              <a target="_self" href="https://scholar.google.com.br/citations?user=d2ij4wUAAAAJ&hl" >Google Scholar</a>
            </li>
            <li>
              <a target="_self" href="https://orcid.org/0000-0003-2143-9701" >ORCID</a>
            </li>
            <li>
              <a target="_self" href="https://publons.com/researcher/2414076/antonio-victor-c-coelho" >Researcher ID/Publons</a>
            </li>
        </ul>
      </nav>

      <ul class="social">
          <li>
            <a  class="sc-github" href="https://github.com/antoniocampos13/portfolio" target="_blank">
              <i class="fab fa-github"></i>
            </a>
          </li>
          <li>
            <a  class="sc-linkedin" href="https://www.linkedin.com/in/antonio-coelho-9aa338164" target="_blank">
              <i class="fab fa-linkedin"></i>
            </a>
          </li>
      </ul>
    </div>

  </aside>
  <main>

    <nav>
      <a href="https://antoniocampos13.github.io">Home</a>

      <a href="/archives.html">Archives</a>
      <a href="/categories.html">Categories</a>
      <a href="/tags.html">Tags</a>

      <a href="https://antoniocampos13.github.io/feeds/all.atom.xml">Atom</a>

      <a href="https://antoniocampos13.github.io/feeds/all.rss.xml">RSS</a>
    </nav>

<article class="single">
  <header>
      
    <h1 id="fastq-to-annotation-part-1"><span class="caps">FASTQ</span> to Annotation (Part&nbsp;1)</h1>
    <p>
      Posted on Thu 01 October 2020 in <a href="https://antoniocampos13.github.io/category/unix.html">Unix</a>

    </p>
  </header>


  <div>
    <h2>Introduction</h2>
<p>In my <a href="https://antoniocampos13.github.io/setting-up-your-unix-computer-for-bioinformatics-analysis.html">previous post</a>, I showed how to configure an Ubuntu system to install Bioinformatics&nbsp;programs.</p>
<p>Now, using the environment I created, I will demonstrate a bash script, <code>FastQ_to_Annotation.sh</code> that takes next generation sequencing (<span class="caps">NGS</span>) raw reads from human whole genome sequencing as input and produces variant annotation as output. Variant annotation is the process of identifying genetic variants in some genomic <span class="caps">DNA</span> sample, and assess, for example, if any of the found variants have any effect on phenotype, such as increased susceptibility to certain&nbsp;diseases.</p>
<p>This demonstration will be separated in four parts. Here in the first part, I will show how to search for <span class="caps">NGS</span> projects deposited in <a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (<span class="caps">NCBI</span>) databases</a> from which I can download sequencing reads later to use with the&nbsp;script.</p>
<h2>Using <span class="caps">NCBI</span>&#8217;s entrez-direct (EDirect) to retrieve <span class="caps">FASTQ</span>&nbsp;files</h2>
<p>I open my Unix terminal and activate the <code>bioenv</code> environment:</p>
<div class="highlight"><pre><span></span><code>conda activate bioenv
</code></pre></div>

<p>Now I use the <code>EDirect</code> <code>esearch</code> command to search <span class="caps">NCBI</span>&#8217;s databases. I must provide a database using the flag <code>-db</code>. Check the available databases <a href="https://www.ncbi.nlm.nih.gov/books/NBK25497/table/chapter2.T._entrez_unique_identifiers_ui/?report=objectonly">here</a>.</p>
<p>I will search the <code>biproject</code> database because it contains metadata from projects dealing with high-throughput genome sequencing, transcriptome expression analysis and so on. I must use the <code>-query</code> flag to provide keywords for search. In this example, I will search for studies dealing with <strong>vorinostat</strong>, a medicine that is have been used in experimental <span class="caps">HIV</span>-1 latency reversal, or &#8220;shock-and-kill&#8221;&nbsp;treatments.</p>
<p>Remember to use single quotes (&#8221;) enclosing the query, especially if it has several&nbsp;words.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># It is just the beginning... (1/4)</span>

<span class="c1"># Let&#39;s create a folder to organize our files inside</span>
mkdir demo
<span class="nb">cd</span> demo

esearch -db bioproject -query <span class="s1">&#39;vorinostat&#39;</span>
</code></pre></div>

<p>The output is just a <code>XML</code> summary including, among other things, the number of results&nbsp;retrieved:</p>
<div class="highlight"><pre><span></span><code><span class="nt">&lt;ENTREZ_DIRECT&gt;</span>
  <span class="nt">&lt;Db&gt;</span>bioproject<span class="nt">&lt;/Db&gt;</span>
  <span class="nt">&lt;WebEnv&gt;</span>MCID_5f7726730525f301023dc947<span class="nt">&lt;/WebEnv&gt;</span>
  <span class="nt">&lt;QueryKey&gt;</span>1<span class="nt">&lt;/QueryKey&gt;</span>
  <span class="nt">&lt;Count&gt;</span>61<span class="nt">&lt;/Count&gt;</span>
  <span class="nt">&lt;Step&gt;</span>1<span class="nt">&lt;/Step&gt;</span>
<span class="nt">&lt;/ENTREZ_DIRECT&gt;</span>
</code></pre></div>

<p>In this case, the query resulted in 61 results (check the <code>&lt;count&gt;</code> tag). Thus, I will add more commands to retrieve the actual query results. I will pipe, i.e. transfer, the results of the query to the another command &#8212; <code>efetch</code> &#8212; that will do this work for me. This is the pipe symbol: <code>|</code>.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># ... not there yet ... (2/4)</span>
esearch -db bioproject -query <span class="s1">&#39;vorinostat&#39;</span> <span class="p">|</span> efetch -format native -mode xml
</code></pre></div>

<p>The output is in <code>XML</code> format, and it is unfortunately not very much human-readable. Here is a print screen showing the first result. Notice how the record is contained within a <code>&lt;/DocumentSummary&gt;</code> node:</p>
<p><img alt="esearch vorinostat XML result" src="https://antoniocampos13.github.io/images/esearch_vorinostat_xml_results.PNG"></p>
<p>Thus, I will once again pipe the results, this time to <code>xtract</code> command. As its name implies, it extracts information from the <code>XML</code> and formats into a tab-separated format that is easier to understand. I must input the flag <code>-pattern</code> with the part of the <code>XML</code> files that contains the desired information, which are <code>elements</code>. In this example, I will search inside the <code>DocumentSummary</code> for <code>ArchiveID@accession</code> (project unique accession number), <code>ID</code> (an auxiliary <span class="caps">ID</span> code to search for samples of said project), <code>Title</code>(the title of the project),  <code>Description</code> (normally an abstract of the project) and <code>Reference</code> (a list of project-related papers in PubMed ids &#8212; PMIDs, if available). Note that I am separating each argument with spaces, no quotes are necessary in this part of the&nbsp;command.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># ...almost there ... (3/4)</span>
esearch -db bioproject -query <span class="s1">&#39;vorinostat&#39;</span> <span class="p">|</span> efetch -format native -mode xml <span class="p">|</span> xtract -pattern DocumentSummary -element ArchiveID@accession ID Title Description Reference
</code></pre></div>

<p>Here is the tab-separated output of the same record displayed on my&nbsp;terminal:</p>
<p><img alt="esearch vorinostat xtract result" src="https://antoniocampos13.github.io/images/esearch_vorinostat_xtract_results.PNG"></p>
<p>Lastly, I will add a final command to transfer to a local text file <code>vorinostat_projects.txt</code> that will be saved in the current working directory. Note that if you have a identically-named file in the working directory, it will be overwritten, so be&nbsp;careful.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Finally there! (4/4)</span>
esearch -db bioproject -query <span class="s1">&#39;vorinostat&#39;</span> <span class="p">|</span> efetch -format native -mode xml <span class="p">|</span> xtract -pattern DocumentSummary -element ArchiveID@accession ID Reference Title Description &gt; vorinostat_projects.txt
</code></pre></div>

<p>The <code>vorinostat_projects.txt</code> file can then be imported into spreadsheets to make easier to organize and read the&nbsp;results.</p>
<h2>Refining the&nbsp;search</h2>
<p>The command above is a very basic one to search <span class="caps">NCBI</span> databases via <code>EDirect</code>. I can create more elaborate queries by adding other keywords and filtering results. <span class="caps">NCBI</span>&#8217;s search engines have several parameters. I advise you go to any advanced search page on the <span class="caps">NCBI</span> website to look for the available&nbsp;parameters.</p>
<p>Using <a href="https://www.ncbi.nlm.nih.gov/bioproject/"><code>BioProject</code> database</a> as example again, click on <em>Advanced</em> to go the query&nbsp;constructor:</p>
<p><img alt="BioProject search box" src="https://antoniocampos13.github.io/images/bioproject_start.PNG"></p>
<p>Using the <strong>BioProject Advanced Search Builder</strong>, I will refine our search. I wish to include only projects that with samples deposited on <a href="https://www.ncbi.nlm.nih.gov/sra">Sequence Read Archive (<span class="caps">SRA</span>)</a>, from human samples and that investigated genetic variation. I input all of this into the search&nbsp;boxes:</p>
<p><img alt="Refining our search" src="https://antoniocampos13.github.io/images/vorinostat_refined.PNG"></p>
<p>Note that clicking in <code>Show index list</code> will provide a list of curated terms. I used them to filter for &#8220;bioproject sra&#8221; and &#8220;variation&#8221; projects. To filter for organism, it is easier: I simply selected the Organism on the drop-down list on the left of the search box. Finally, I connected all keywords with the <code>AND</code> Boolean constructor, resulting on the&nbsp;query:</p>
<p><code>(((vorinostat) AND "bioproject sra"[Filter]) AND Homo sapiens[Organism]) AND "variation"[Filter]</code></p>
<p>You could continue the search on the website, of course, but let&#8217;s go back to the terminal and continue from&nbsp;there:</p>
<div class="highlight"><pre><span></span><code>esearch -db bioproject -query <span class="s1">&#39;(((vorinostat) AND &quot;bioproject sra&quot;[Filter]) AND Homo sapiens[Organism]) AND &quot;variation&quot;[Filter]&#39;</span>  <span class="p">|</span> efetch -format native -mode xml <span class="p">|</span> xtract -pattern DocumentSummary -element ArchiveID@accession ID Reference Title Description  &gt; vorinostat_refined.txt
</code></pre></div>

<p>Remember: single quotes enclosing the query. Turns out that this refined search was rather restrictive: it resulted in a single record. Checking the <code>vorinostat_refined.txt</code> I see in the abstract that the project dealt with samples from patients with melanoma. One of the last sentences says: &#8220;<span class="caps">DNA</span> Seq data: biopsy samples from patients pre- and post- treated with Vorinostat; check mutations related to MAPKi-resistance&#8221; (MAPKi: Mitogen Activated Protein Kinase inhibitors). Although I had <span class="caps">HIV</span>-1-related projects in mind, that&#8217;s fine for now, since it is suitable to <code>FastQ_to_Annotation.sh</code> script: identify and annotate genetic&nbsp;variation.</p>
<p>Then, I take note of the project <span class="caps">ID</span>: <code>PRJNA436005</code>. I will use it to retrieve reads from this project by searching the <span class="caps">SRA</span> with&nbsp;it.</p>
<h2>Conclusion of Part&nbsp;1</h2>
<p>In this part I showed how&nbsp;to:</p>
<ul>
<li>search <span class="caps">NCBI</span>&#8217;s databases, (especially&nbsp;BioProject);</li>
<li>refine&nbsp;searches;</li>
<li>save search results into local, human-readable text&nbsp;files.</li>
</ul>
<p>Now I need to use the information gathered here to download read sets in <code>FASTQ</code> format.</p>
<p><em><a href="https://antoniocampos13.github.io/fastq-to-annotation-part-2">Go to <span class="caps">FASTQ</span> to Annotation (Part&nbsp;2)</a></em></p>
<h2>References</h2>
<p><a href="https://antoniocampos13.github.io/setting-up-your-unix-computer-for-bioinformatics-analysis.html">Setting Up Your Unix Computer for Bioinformatics&nbsp;Analysis</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology&nbsp;Information</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/books/NBK25497/table/chapter2.T._entrez_unique_identifiers_ui/?report=objectonly">Entrez Unique Identifiers (UIDs) for selected&nbsp;databases</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/bioproject/">Home - BioProject - <span class="caps">NCBI</span></a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/sra">Home - <span class="caps">SRA</span> - <span class="caps">NCBI</span></a></p>
  </div>
  <div class="tag-cloud">
    <p>
      <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>
      <a href="https://antoniocampos13.github.io/tag/genomic-variation.html">genomic variation</a>
      <a href="https://antoniocampos13.github.io/tag/entrez-direct.html">entrez-direct</a>
      <a href="https://antoniocampos13.github.io/tag/edirect.html">EDirect</a>
    </p>
  </div>





</article>

    <footer>
<p>
  &copy; 2020 Antonio Victor Campos Coelho - This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>
</p>
<p>
Built with <a href="http://getpelican.com" target="_blank">Pelican</a> using <a href="http://bit.ly/flex-pelican" target="_blank">Flex</a> theme
</p><p>
  <a rel="license"
     href="http://creativecommons.org/licenses/by-sa/4.0/"
     target="_blank">
    <img alt="Creative Commons License"
         title="Creative Commons License"
         style="border-width:0"
           src="https://i.creativecommons.org/l/by-sa/4.0/80x15.png"
         width="80"
         height="15"/>
  </a>
</p>    </footer>
  </main>




<script type="application/ld+json">
{
  "@context" : "http://schema.org",
  "@type" : "Blog",
  "name": " Antonio's Portfolio ",
  "url" : "https://antoniocampos13.github.io",
  "image": "https://avatars.githubusercontent.com/antoniocampos13",
  "description": "Data Science Portfolio by Antonio Victor Campos Coelho"
}
</script>


</body>
</html>