
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <meta name="HandheldFriendly" content="True" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta name="robots" content="" />

  <link href="https://fonts.googleapis.com/css2?family=Source+Code+Pro:ital,wght@0,400;0,700;1,400&family=Source+Sans+Pro:ital,wght@0,300;0,400;0,700;1,400&display=swap" rel="stylesheet">

    <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/stylesheet/style.min.css">


    <link id="pygments-light-theme" rel="stylesheet" type="text/css"
          href="https://antoniocampos13.github.io/theme/pygments/vs.min.css">


  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/fontawesome.css">
  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/brands.css">
  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/solid.css">


    <link href="https://antoniocampos13.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="Antonio's Portfolio Atom">

    <link href="https://antoniocampos13.github.io/feeds/all.rss.xml" type="application/rss+xml" rel="alternate" title="Antonio's Portfolio RSS">


  

    <!-- Chrome, Firefox OS and Opera -->
    <meta name="theme-color" content="#333333">
    <!-- Windows Phone -->
    <meta name="msapplication-navbutton-color" content="#333333">
    <!-- iOS Safari -->
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    <!-- Microsoft EDGE -->
    <meta name="msapplication-TileColor" content="#333333">

 

<meta name="author" content="Antonio Victor Campos Coelho" />
<meta name="description" content="Introduction In a previous post, I showed how to configure an Ubuntu system to install Bioinformatics programs. Now, using the environment I created, I will demonstrate a bash script, FastQ_to_Annotation.sh that takes next generation sequencing (NGS) raw reads from human whole genome sequencing as input and produces …" />
<meta name="keywords" content="Bioinformatics, genomic variation, entrez-direct, EDirect">


  <meta property="og:site_name" content="Antonio's Portfolio"/>
  <meta property="og:title" content="FASTQ to Annotation (Part 4)"/>
  <meta property="og:description" content="Introduction In a previous post, I showed how to configure an Ubuntu system to install Bioinformatics programs. Now, using the environment I created, I will demonstrate a bash script, FastQ_to_Annotation.sh that takes next generation sequencing (NGS) raw reads from human whole genome sequencing as input and produces …"/>
  <meta property="og:locale" content="en_US"/>
  <meta property="og:url" content="https://antoniocampos13.github.io/fastq-to-annotation-part-4.html"/>
  <meta property="og:type" content="article"/>
  <meta property="article:published_time" content="2020-10-06 18:00:00-03:00"/>
  <meta property="article:modified_time" content=""/>
  <meta property="article:author" content="https://antoniocampos13.github.io/author/antonio-victor-campos-coelho.html">
  <meta property="article:section" content="Unix"/>
  <meta property="article:tag" content="Bioinformatics"/>
  <meta property="article:tag" content="genomic variation"/>
  <meta property="article:tag" content="entrez-direct"/>
  <meta property="article:tag" content="EDirect"/>
  <meta property="og:image" content="https://avatars.githubusercontent.com/antoniocampos13">

  <title>Antonio's Portfolio &ndash; FASTQ to Annotation (Part 4)</title>

</head>
<body class="light-theme">
  <aside>
    <div>
      <a href="https://antoniocampos13.github.io/">
        <img src="https://avatars.githubusercontent.com/antoniocampos13" alt="Antonio's Portfolio" title="Antonio's Portfolio">
      </a>

      <h1>
        <a href="https://antoniocampos13.github.io/">Antonio's Portfolio</a>
      </h1>

<p>PhD in Genetics</p>

      <nav>
        <ul class="list">


              <li>
                <a target="_self"
                   href="https://antoniocampos13.github.io/pages/about.html#about">
                  About
                </a>
              </li>
              <li>
                <a target="_self"
                   href="https://antoniocampos13.github.io/pages/contact.html#contact">
                  Contact
                </a>
              </li>

            <li>
              <a target="_self" href="http://lattes.cnpq.br/2986394950644755" >Brazilian Lattes CV</a>
            </li>
            <li>
              <a target="_self" href="https://scholar.google.com.br/citations?user=d2ij4wUAAAAJ&hl" >Google Scholar</a>
            </li>
            <li>
              <a target="_self" href="https://orcid.org/0000-0003-2143-9701" >ORCID</a>
            </li>
            <li>
              <a target="_self" href="https://publons.com/researcher/2414076/antonio-victor-c-coelho" >Researcher ID/Publons</a>
            </li>
        </ul>
      </nav>

      <ul class="social">
          <li>
            <a  class="sc-github" href="https://github.com/antoniocampos13/portfolio" target="_blank">
              <i class="fab fa-github"></i>
            </a>
          </li>
          <li>
            <a  class="sc-linkedin" href="https://www.linkedin.com/in/antonio-coelho-9aa338164" target="_blank">
              <i class="fab fa-linkedin"></i>
            </a>
          </li>
      </ul>
    </div>

  </aside>
  <main>

    <nav>
      <a href="https://antoniocampos13.github.io/">Home</a>

      <a href="/archives.html">Archives</a>
      <a href="/categories.html">Categories</a>
      <a href="/tags.html">Tags</a>

      <a href="https://antoniocampos13.github.io/feeds/all.atom.xml">Atom</a>

      <a href="https://antoniocampos13.github.io/feeds/all.rss.xml">RSS</a>
    </nav>

<article class="single">
  <header>
      
    <h1 id="fastq-to-annotation-part-4"><span class="caps">FASTQ</span> to Annotation (Part&nbsp;4)</h1>
    <p>
      Posted on Tue 06 October 2020 in <a href="https://antoniocampos13.github.io/category/unix.html">Unix</a>

    </p>
  </header>


  <div>
    <h2>Introduction</h2>
<p><em>In a <a href="https://antoniocampos13.github.io/setting-up-your-unix-computer-for-bioinformatics-analysis.html">previous post</a>, I showed how to configure an Ubuntu system to install Bioinformatics&nbsp;programs.</em></p>
<p><em>Now, using the environment I created, I will demonstrate a bash script, <code>FastQ_to_Annotation.sh</code> that takes next generation sequencing (<span class="caps">NGS</span>) raw reads from human whole genome sequencing as input and produces variant annotation as output. Variant annotation is the process of identifying genetic variants in some genomic <span class="caps">DNA</span> sample, and assess, for example, if any of the found variants have any effect on phenotype, such as increased susceptibility to certain&nbsp;diseases.</em></p>
<p><em>In the <a href="https://antoniocampos13.github.io/fastq-to-annotation-part-1">first part</a>, I showed how to search for <span class="caps">NGS</span> projects deposited in <a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (<span class="caps">NCBI</span>) databases</a> from which I can download sequencing reads later to use with the&nbsp;script.</em></p>
<p><em>In the <a href="https://antoniocampos13.github.io/fastq-to-annotation-part-2">second part</a>, I showed how to retrieve raw genome sequencing reads in the form of <code>FASTQ</code> files, which are deposited in <a href="https://www.ncbi.nlm.nih.gov/sra"><span class="caps">SRA</span></a>.</em></p>
<p><em>In the <a href="https://antoniocampos13.github.io/fastq-to-annotation-part-3">third part</a>, I made the final preparations for the <code>FastQ_to_Annotation.sh</code> script demonstration using the <code>FASTQ</code> files obtained in the second&nbsp;part.</em></p>
<p>Here in the fourth and final part, I finally can summarize the inner workings of the <code>FastQ_to_Annotation.sh</code> script.</p>
<h2>FastQ_to_Annotation.sh&nbsp;parameters</h2>
<p><strong>Activate</strong> your miniconda environment if needed and go to your <code>demo</code> folder. Make sure you have the <code>FASTQ</code> files and a <code>refs</code> folder with the human genome <code>FASTA</code> files and the other various supporting&nbsp;files.</p>
<p>The script needs 10 command line parameters to work correctly. They&nbsp;are:</p>
<ul>
<li>Mate-pair <span class="caps">FASTQ</span> files name root (without extension) (absolute file&nbsp;path)</li>
<li>Reference genome <span class="caps">FASTA</span> (absolute file&nbsp;path)</li>
<li><span class="caps">BED</span> or <span class="caps">GFF</span> file (absolute file&nbsp;path)</li>
<li>Minimum quality for bases at read ends, below which bases will be cut (integer - default:&nbsp;20)</li>
<li>Minimum allowed read length (integer - default:&nbsp;20)</li>
<li>Adaptor for trimming off read ends (&#8216;illumina&#8217; / &#8216;nextera&#8217; /&nbsp;&#8216;small_rna&#8217;)</li>
<li>Minimum read depth for calling a variant (integer - default:&nbsp;3)</li>
<li>Minimum allowed mapping quality (integer - default:&nbsp;0)</li>
<li>Stringency for calling variants (&#8216;relaxed&#8217; / &#8216;normal&#8217;) (relaxed uses &#8212;pval-threshold 1.0 with BCFtools&nbsp;call)</li>
<li>User identification for logging&nbsp;(alphanumeric)</li>
</ul>
<p>For this example, use the following&nbsp;values:</p>
<ul>
<li><code>SRR6784104</code></li>
<li><code>refs/GCA_000001405.15_GRCh38_no_alt_plus_hs38d1_analysis_set.fna.gz</code></li>
<li><code>refs/GCA_000001405.15_GRCh38_no_alt_plus_hs38d1_analysis_set.fna.bed</code></li>
<li><code>20</code></li>
<li><code>20</code></li>
<li><code>illumina</code></li>
<li><code>3</code></li>
<li><code>0</code></li>
<li><code>normal</code></li>
<li>Your name (do not use&nbsp;spaces)</li>
</ul>
<p>Since the names of the compressed human genome <code>FASTA</code> file is big, you can rename it, or create an alias in the command line to simplify the&nbsp;command:</p>
<div class="highlight"><pre><span></span><code><span class="nv">REF</span><span class="o">=</span>refs/GCA_000001405.15_GRCh38_no_alt_plus_hs38d1_analysis_set.fna.gz
<span class="nv">BED</span><span class="o">=</span>refs/GCA_000001405.15_GRCh38_no_alt_plus_hs38d1_analysis_set.fna.bed
</code></pre></div>

<p>Then, joining everything&nbsp;together:</p>
<div class="highlight"><pre><span></span><code>./FastQ_to_Annotation.sh SRR6784104 <span class="nv">$REF</span> <span class="nv">$BED</span> <span class="m">20</span> <span class="m">20</span> illumina <span class="m">3</span> <span class="m">0</span> normal antonio
</code></pre></div>

<h2>Pipeline&nbsp;steps</h2>
<p>The script will check if all parameters are adequate and then run the core pipeline, which proceeds in an 8-step&nbsp;process:</p>
<ol>
<li>
<p>Adaptor and read quality trimming: uses <a href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">Trim Galore!</a>, <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> and <a href="https://github.com/marcelm/cutadapt/">Cutadapt</a> programs. They remove adaptor sequence from reads and discards low-quality reads so they do not interfere with the second step, alignment. Outputs the trimmed <code>FASTQ</code> files, text and <code>HTML</code> reports of the trimming&nbsp;results.</p>
</li>
<li>
<p>Alignment: uses <code>bwa mem</code> command (<a href="https://academic.oup.com/bioinformatics/article/25/14/1754/225615">Li <span class="amp">&amp;</span> Durbin, 2009</a>). <code>bwa</code> is a widely-used program to align short reads into genomes, so we can pinpoint where in the genome the identified variants are located. Takes the trimmed <code>FASTQ</code> files, the reference <code>FASTA</code> file and produces an aligned <span class="caps">SAM</span>&nbsp;file.</p>
</li>
<li>
<p>Marking and removing <span class="caps">PCR</span> duplicates: uses Picard (Broad Institute of <span class="caps">MIT</span> and Harvard) and SAMtools <a href="https://academic.oup.com/bioinformatics/article/25/16/2078/204688">(Li et al., 2009)</a>. This is another cleanup step. It takes the aligned <span class="caps">SAM</span> file and produces an aligned sorted <span class="caps">BAM</span> file with duplicated reads removed. <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-016-1097-3">Ebbert et al.</a> define <span class="caps">PCR</span> duplicates as: &#8220;&#8230;sequence reads that result from sequencing two or more copies of the exact same <span class="caps">DNA</span> fragment, which, at worst, may contain erroneous mutations introduced during <span class="caps">PCR</span> amplification, or, at the very least, make the occurrence of the allele(s) sequenced in duplicates appear proportionately more often than it should compared to the other allele (assuming a non-haploid&nbsp;organism)&#8221;.</p>
</li>
<li>
<p>Remove low mapping quality reads: uses SAMtools (Li et al., 2009). Reads falling in repetitive regions usually get very low mapping quality, so we remove it to reduce noise during variant call. Takes the aligned sorted <span class="caps">BAM</span> file with duplicated reads removed and removes low mapping quality&nbsp;reads.</p>
</li>
<li>
<p>Quality control (<span class="caps">QC</span>): uses SAMtools (Li et al., 2009), BEDTools <a href="https://academic.oup.com/bioinformatics/article/26/6/841/244688">(Quinlan <span class="amp">&amp;</span> Hall, 2010)</a>. Quantifies the removed off-target reads, the sequencing reads that do not align to the target genome and calculates the mean depth of read coverage in the genome. Takes in the <span class="caps">BAM</span> file generated in the previous&nbsp;step.</p>
</li>
<li>
<p>Downsampling/random read sampling: uses <a href="https://broadinstitute.github.io/picard/">Picard</a> (Broad Institute of <span class="caps">MIT</span> and Harvard). This step takes the cleaned-up aligned sorted <span class="caps">BAM</span> file generated by the previous steps and splits into 3 &#8216;sub-BAMs&#8217; of random reads sorted with probabilities of 75%, 50%, and&nbsp;25%.</p>
</li>
<li>
<p>Variant calling: uses SAMtools/BCFtools (Li et al., 2009). This step identifies genetic variation present in the sample reads. It takes on all 4 <span class="caps">BAM</span> files, after which a consensus <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3137218/">Variant Call Format (<span class="caps">VCF</span>)</a> file is&nbsp;produced.</p>
</li>
<li>
<p>Annotation: uses Variant Effect Predictor <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0974-4">(McLaren et al., 2016)</a>. It takes the list of variants compiled in the consensus <span class="caps">VCF</span> file and annotates them, identifying possible phenotypic effects. Outputs text and html summary files with the results. Check <a href="https://www.ensembl.org/info/docs/tools/vep/script/vep_other.html"><span class="caps">VEP</span>&#8217;s documentation</a> if you want to customize the annotation options in the&nbsp;script.</p>
</li>
</ol>
<h2>Output</h2>
<p>Once the script is running you will see several files being generated. Once the script finishes, the files will be neatly organized in a folder <code>prefix_results</code>, where <code>prefix</code> is the name root of the <code>FASTQ</code> files:</p>
<p><img alt="demo results 1" src="https://antoniocampos13.github.io/images/demo_results_1.PNG"></p>
<p>Open the folder and check that there are some files and three subfolders. These subfolders hold all intermediate files generated by the script (<code>.sam</code>, <code>.bam</code> and many others). <code>trimmed_files</code> folder hold the trimmed <code>FASTQ</code> files alongside Trim Galore!&#8217;s reports (step 1). <code>alignment_files</code> hold intermediate files generated by steps 2 trough 5. <code>variant_call_files</code> hold intermediate files generated by steps 7 through&nbsp;8.</p>
<p><img alt="demo results 2" src="https://antoniocampos13.github.io/images/demo_results_2.PNG"></p>
<p>Let&#8217;s focus the attention on the other five&nbsp;files:</p>
<ul>
<li>
<p>Master_Log.txt and Pipeline_Log.txt files: logs from the script operations. The first one has a copy of all commands issued by the script. The second one is more concise; it summarizes input parameters alongside date and time each step in the scripted started. Check these files if any errors occur to identify what went&nbsp;wrong.</p>
</li>
<li>
<p>Final.vcf: a <span class="caps">VCF</span> file containing all variants identified in the sample. It contains chromosome position of the variants, alleles and other&nbsp;information.</p>
</li>
<li>
<p>AnnotationVEP.txt and AnnotationVEP.html: outputs of annotation by Ensembl&#8217;s <span class="caps">VEP</span>. The text file is tab-separated file listing the called variants and their characteristics (more on that later). The <code>HTML</code> file contains a summarized quantification of the variants&nbsp;characteristics.</p>
</li>
</ul>
<p>Open the <code>SRR6784104_AnnotationVEP.txt</code> file into a spreadsheet to make the visualization easier. You will see there is a header with several definitions/abbreviations for the information contained in the file. Scroll down until you found a table-like&nbsp;part.</p>
<p>In this table part, there is several important information that is interesting to check. Some of the columns I like to&nbsp;assess:</p>
<ul>
<li><code>#Uploaded_variation</code>: an identifier of each&nbsp;variation;</li>
<li><code>Location</code>: chromosome and position of the&nbsp;variation;</li>
<li><code>Allele</code>: particular nucleotide configuration found in determined position in the&nbsp;sample;</li>
<li><code>Gene</code>: if the variant is located within a gene, its unique RefSeq gene <span class="caps">ID</span> (an integer) will be&nbsp;there;</li>
<li><code>Feature</code>: if the variant is located within a gene, a unique RefSeq accession code of the gene sequence will be&nbsp;there;</li>
<li><code>Consequence</code>: I found this column weirdly-named, because it reflects more the overall location of the variant than a molecular consequence as the name implies. For example, it will indicate that the variant is a <code>missense_variant</code>, an <code>intron_variant</code>, <code>regulatory_region_variant</code> and so&nbsp;on;</li>
<li><code>Protein_position</code>, <code>Amino_acids</code>, <code>Codons</code>: if missense or synonym, information about amino acids changes and position on the protein will be in these&nbsp;columns;</li>
<li><code>Existing_variation</code>: if variation was already previously identified in other samples, the RefSeq (starting with <code>rs</code>) or other identifier will be there. RefSeq-identified variants can be found in <a href="https://www.ncbi.nlm.nih.gov/snp/"><span class="caps">NCBI</span>&#8217;s dbSNP</a>;</li>
<li><code>IMPACT</code>: the variant&#8217;s impact on phenotype (<span class="caps">LOW</span>, <span class="caps">MODIFIER</span>, <span class="caps">HIGH</span>);</li>
<li><code>VARIANT_CLASS</code>: the class of the variant. <span class="caps">SNV</span> (single nucleotide variation, the same as single nucleotied polymorphism &#8212; <span class="caps">SNP</span>), insertions and deletions are the most&nbsp;common;</li>
<li><code>SYMBOL</code>: the official symbol (abbreviation) of the gene&nbsp;name;</li>
<li><code>BIOTYPE</code>: if the variant is located within a gene, the gene function. For example: protein_coding, lncRNA, miRNA, and so&nbsp;on;</li>
<li><code>SIFT</code> and <code>PolyPhen</code>: named after the tools that predict whether an amino acid substitution affects protein function and structure of a human&nbsp;protein;</li>
<li>Columns prefixed with <code>AF</code>: contain the allelic frequency of a given variant in some <a href="https://www.internationalgenome.org/category/population/">global populations</a>. For example, <code>AFR</code>: African,<code>AMR</code>: Ad Mixed American, <code>EAS</code>: East Asian, <code>SAS</code>: South&nbsp;Asian;</li>
<li><code>CLIN_SIG</code>: a short sentence stating the clinical significance (if available) of the&nbsp;variant;</li>
<li><code>PUBMED</code>: a list of PubMed IDs of references citing the variation (if&nbsp;available).</li>
</ul>
<p>The <code>AnnotationVEP.html</code> file contains a collection of graphical representations of several characteristics of the detected variants. See below some of them. Notice that your results will be different from these figures, since I used a different set of <code>FASTQ</code> files and reference&nbsp;files.</p>
<p><img alt="demo results 3" src="https://antoniocampos13.github.io/images/demo_results_3.PNG"></p>
<h2>Conclusion of Part&nbsp;4</h2>
<p>In this part I&nbsp;showed:</p>
<ul>
<li>How to use the <code>FastQ_to_Annotation.sh</code> script;</li>
<li>Summarized the steps performed by the&nbsp;script;</li>
<li>Summarized the principal results output by the&nbsp;script.</li>
</ul>
<p>Therefore, I finished all the steps I followed to prepare the system for Bioinformatics analysis, gather the necessary files and apply them to obtain annotations from human genome <span class="caps">NGS</span> reads&nbsp;samples.</p>
<p>Subscribe to my <a href="https://antoniocampos13.github.io/feeds/all.rss.xml">rss feed</a> or <a href="https://antoniocampos13.github.io/feeds/all.atom.xml">Atom feed</a> to keep updated whenever I post new&nbsp;protocols.</p>
<p><em><a href="https://antoniocampos13.github.io/fastq-to-annotation-part-1">Go back to <span class="caps">FASTQ</span> to Annotation (Part&nbsp;1)</a></em></p>
<p><em><a href="https://antoniocampos13.github.io/fastq-to-annotation-part-2">Go back to <span class="caps">FASTQ</span> to Annotation (Part&nbsp;2)</a></em></p>
<p><em><a href="https://antoniocampos13.github.io/fastq-to-annotation-part-3">Go back to <span class="caps">FASTQ</span> to Annotation (Part&nbsp;3)</a></em></p>
<h2>References</h2>
<p><a href="https://antoniocampos13.github.io/setting-up-your-unix-computer-for-bioinformatics-analysis.html">Setting Up Your Unix Computer for Bioinformatics&nbsp;Analysis</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology&nbsp;Information</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/sra">Home - <span class="caps">SRA</span> - <span class="caps">NCBI</span></a></p>
<p><a href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">Babraham Bioinformatics - Trim&nbsp;Galore!</a></p>
<p><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">Babraham Bioinformatics - FastQC A Quality Control tool for High Throughput Sequence&nbsp;Data</a></p>
<p><a href="https://github.com/marcelm/cutadapt/">marcelm/cutadapt</a></p>
<p><a href="https://academic.oup.com/bioinformatics/article/25/14/1754/225615">Fast and accurate short read alignment with Burrows–Wheeler&nbsp;transform</a></p>
<p><a href="https://academic.oup.com/bioinformatics/article/25/16/2078/204688">Sequence Alignment/Map format and&nbsp;SAMtools</a></p>
<p><a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-016-1097-3">Evaluating the necessity of <span class="caps">PCR</span> duplicate removal from next-generation sequencing data and a comparison of&nbsp;approaches</a></p>
<p><a href="https://academic.oup.com/bioinformatics/article/26/6/841/244688">BEDTools: a flexible suite of utilities for comparing genomic&nbsp;features</a></p>
<p><a href="https://broadinstitute.github.io/picard/">Picard Tools - By Broad&nbsp;Institute</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3137218/">The variant call format and&nbsp;VCFtools</a></p>
<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0974-4">The Ensembl Variant Effect&nbsp;Predictor</a></p>
<p><a href="https://www.ensembl.org/info/docs/tools/vep/script/vep_other.html">Other&nbsp;information</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/snp/">Home - <span class="caps">SNP</span> - <span class="caps">NCBI</span></a></p>
<p><a href="https://www.internationalgenome.org/category/population/">Population | 1000&nbsp;Genomes</a></p>
  </div>
  <div class="tag-cloud">
    <p>
      <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>
      <a href="https://antoniocampos13.github.io/tag/genomic-variation.html">genomic variation</a>
      <a href="https://antoniocampos13.github.io/tag/entrez-direct.html">entrez-direct</a>
      <a href="https://antoniocampos13.github.io/tag/edirect.html">EDirect</a>
    </p>
  </div>





</article>

    <footer>
<p>
  &copy; 2020 Antonio Victor Campos Coelho - This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/deed.en_US" target="_blank">Creative Commons Attribution-ShareAlike</a>
</p>
<p>
Built with <a href="http://getpelican.com" target="_blank">Pelican</a> using <a href="http://bit.ly/flex-pelican" target="_blank">Flex</a> theme
</p><p>
  <a rel="license"
     href="http://creativecommons.org/licenses/by-sa/4.0/"
     target="_blank">
    <img alt="Creative Commons License"
         title="Creative Commons License"
         style="border-width:0"
           src="https://i.creativecommons.org/l/by-sa/4.0/80x15.png"
         width="80"
         height="15"/>
  </a>
</p>    </footer>
  </main>




<script type="application/ld+json">
{
  "@context" : "http://schema.org",
  "@type" : "Blog",
  "name": " Antonio's Portfolio ",
  "url" : "https://antoniocampos13.github.io",
  "image": "https://avatars.githubusercontent.com/antoniocampos13",
  "description": "Data Science Portfolio by Antonio Victor Campos Coelho"
}
</script>

</body>
</html>