
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <meta name="HandheldFriendly" content="True" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta name="robots" content="" />

  <link href="https://fonts.googleapis.com/css2?family=Source+Code+Pro:ital,wght@0,400;0,700;1,400&family=Source+Sans+Pro:ital,wght@0,300;0,400;0,700;1,400&display=swap" rel="stylesheet">

    <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/stylesheet/style.min.css">


    <link id="pygments-light-theme" rel="stylesheet" type="text/css"
          href="https://antoniocampos13.github.io/theme/pygments/friendly.min.css">


  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/fontawesome.css">
  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/brands.css">
  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/solid.css">


    <link href="https://antoniocampos13.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="Antonio's Portfolio Atom">

    <link href="https://antoniocampos13.github.io/feeds/all.rss.xml" type="application/rss+xml" rel="alternate" title="Antonio's Portfolio RSS">



    <!-- Chrome, Firefox OS and Opera -->
    <meta name="theme-color" content="#333333">
    <!-- Windows Phone -->
    <meta name="msapplication-navbutton-color" content="#333333">
    <!-- iOS Safari -->
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    <!-- Microsoft EDGE -->
    <meta name="msapplication-TileColor" content="#333333">

<meta name="author" content="Antonio Victor Campos Coelho" />
<meta name="description" content="Introduction In a previous post, I showed how to configure an Ubuntu system to install Bioinformatics programs. Now, using the environment I created, I will demonstrate a bash script, FastQ_to_Annotation.sh that takes next generation sequencing (NGS) raw reads from human whole genome sequencing as input and produces variant annotation …" />
<meta name="keywords" content="Bioinformatics, genomic variation, entrez-direct, EDirect">


<meta property="og:site_name" content="Antonio's Portfolio"/>
<meta property="og:title" content="FASTQ to Annotation (Part 3)"/>
<meta property="og:description" content="Introduction In a previous post, I showed how to configure an Ubuntu system to install Bioinformatics programs. Now, using the environment I created, I will demonstrate a bash script, FastQ_to_Annotation.sh that takes next generation sequencing (NGS) raw reads from human whole genome sequencing as input and produces variant annotation …"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:url" content="https://antoniocampos13.github.io/fastq-to-annotation-part-3.html"/>
<meta property="og:type" content="article"/>
<meta property="article:published_time" content="2020-10-05 18:00:00-03:00"/>
<meta property="article:modified_time" content=""/>
<meta property="article:author" content="https://antoniocampos13.github.io/author/antonio-victor-campos-coelho.html">
<meta property="article:section" content="Unix"/>
<meta property="article:tag" content="Bioinformatics"/>
<meta property="article:tag" content="genomic variation"/>
<meta property="article:tag" content="entrez-direct"/>
<meta property="article:tag" content="EDirect"/>
<meta property="og:image" content="https://avatars.githubusercontent.com/antoniocampos13">

  <title>Antonio's Portfolio &ndash; FASTQ to Annotation (Part 3)</title>

</head>
<body class="light-theme">
  <aside>
    <div>
      <a href="https://antoniocampos13.github.io">
        <img src="https://avatars.githubusercontent.com/antoniocampos13" alt="Antonio's Portfolio" title="Antonio's Portfolio">
      </a>

      <h1>
        <a href="https://antoniocampos13.github.io">Antonio's Portfolio</a>
      </h1>

<p>PhD in Genetics</p>

      <nav>
        <ul class="list">


              <li>
                <a target="_self"
                   href="https://antoniocampos13.github.io/pages/about.html#about">
                  About
                </a>
              </li>
              <li>
                <a target="_self"
                   href="https://antoniocampos13.github.io/pages/contact.html#contact">
                  Contact
                </a>
              </li>

            <li>
              <a target="_self" href="http://lattes.cnpq.br/2986394950644755" >Brazilian Lattes CV</a>
            </li>
            <li>
              <a target="_self" href="https://orcid.org/0000-0003-2143-9701" >ORCID</a>
            </li>
            <li>
              <a target="_self" href="https://publons.com/researcher/2414076/antonio-victor-c-coelho" >Researcher ID/Publons</a>
            </li>
        </ul>
      </nav>

      <ul class="social">
          <li>
            <a  class="sc-github" href="https://github.com/antoniocampos13/portfolio" target="_blank">
              <i class="fab fa-github"></i>
            </a>
          </li>
          <li>
            <a  class="sc-linkedin" href="https://www.linkedin.com/in/antonio-coelho-9aa338164" target="_blank">
              <i class="fab fa-linkedin"></i>
            </a>
          </li>
      </ul>
    </div>

  </aside>
  <main>

    <nav>
      <a href="https://antoniocampos13.github.io">Home</a>

      <a href="/archives.html">Archives</a>
      <a href="/categories.html">Categories</a>
      <a href="/tags.html">Tags</a>

      <a href="https://antoniocampos13.github.io/feeds/all.atom.xml">Atom</a>

      <a href="https://antoniocampos13.github.io/feeds/all.rss.xml">RSS</a>
    </nav>

<article class="single">
  <header>
      
    <h1 id="fastq-to-annotation-part-3"><span class="caps">FASTQ</span> to Annotation (Part&nbsp;3)</h1>
    <p>
      Posted on Mon 05 October 2020 in <a href="https://antoniocampos13.github.io/category/unix.html">Unix</a>

    </p>
  </header>


  <div>
    <h2>Introduction</h2>
<p><em>In a <a href="https://antoniocampos13.github.io/setting-up-your-unix-computer-for-bioinformatics-analysis.html">previous post</a>, I showed how to configure an Ubuntu system to install Bioinformatics&nbsp;programs.</em></p>
<p><em>Now, using the environment I created, I will demonstrate a bash script, <code>FastQ_to_Annotation.sh</code> that takes next generation sequencing (<span class="caps">NGS</span>) raw reads from human whole genome sequencing as input and produces variant annotation as output. Variant annotation is the process of identifying genetic variants in some genomic <span class="caps">DNA</span> sample, and assess, for example, if any of the found variants have any effect on phenotype, such as increased susceptibility to certain&nbsp;diseases.</em></p>
<p><em>In the <a href="https://antoniocampos13.github.io/fastq-to-annotation-part-1">first part</a>, I showed how to search for <span class="caps">NGS</span> projects deposited in <a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (<span class="caps">NCBI</span>) databases</a> from which I can download sequencing reads later to use with the&nbsp;script.</em></p>
<p><em>In the <a href="https://antoniocampos13.github.io/fastq-to-annotation-part-2">second part</a>, I showed how to retrieve raw genome sequencing reads in the form of <code>FASTQ</code> files, which are deposited in <a href="https://www.ncbi.nlm.nih.gov/sra"><span class="caps">SRA</span></a>.</em></p>
<p>Here in the third part, I make the final preparations for the <code>FastQ_to_Annotation.sh</code> script demonstration using the <code>FASTQ</code> files obtained in the previous&nbsp;part.</p>
<h2>Final&nbsp;preparations</h2>
<h3>Installing local cache of Ensembl Variant Effect Predictor (<span class="caps">VEP</span>)</h3>
<p>The <a href="https://www.ensembl.org/info/docs/tools/vep/index.html">Ensembl Variant Effect Predictor (<span class="caps">VEP</span>)</a> is the core tool used by the script for the annotation of the effects of any variants present in the sample. It may be used online, but Ensembl strongly recommends users to download and install a local cache of all data deposited in the tool to avoid server overload. I open a terminal, activate the <code>bioenv</code> miniconda environment and execute the commands&nbsp;below:</p>
<p><strong><span class="caps">WARNING</span>: Several gigabytes of data will be downloaded from the internet and installed on your computer. Be sure that you have plenty or unlimited data allowances from your <span class="caps">ISP</span> and sufficient free space on your hard drive before continuing. It will take a while (several minutes to hours) until all the needed processes&nbsp;finish.</strong></p>
<div class="highlight"><pre><span></span><code>vep_install -a cf -s homo_sapiens_refseq -y GRCh38 -c . –CONVERT
</code></pre></div>


<p>Unfortunately, the command above is prone to network and other esoteric errors. If you have any problem, you can try an alternative manner. See&nbsp;below:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Alternative: manually download the compressed cache</span>
wget ftp://ftp.ensembl.org/pub/release-101/variation/indexed_vep_cache/homo_sapiens_refseq_vep_101_GRCh38.tar.gz -P <span class="nv">$HOME</span>/.vep

<span class="c1"># Uncompress the cache</span>
tar -zxf <span class="nv">$HOME</span>/.vep/homo_sapiens_refseq_vep_101_GRCh38.tar.gz -C <span class="nv">$HOME</span>/.vep
</code></pre></div>


<p>After some time (several minutes to some hours depending on the network and the capabilities of your computer) the <span class="caps">VEP</span> cache will be downloaded and installed into a hidden folder in your home folder (<code>$HOME/.vep</code>). Therefore, notice that it is independent of miniconda environments. Thus, it is expected that once installed, this cache will work with any miniconda environment on your computer. You may backup the <code>.vep/</code> folder to avoid downloading the whole thing again (but consider to download newer versions of the cache as they become available&nbsp;though).</p>
<h3>Obtaining human genome reference&nbsp;files</h3>
<p>The annotation process require a collection of reference files. These files will assist us to generate a &#8220;list&#8221; of genetic variants, alongside their possible effects on the phenotype (the <strong>annotation</strong>).</p>
<p>These files&nbsp;are:</p>
<ol>
<li>A <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=BlastHelp"><code>FASTA</code></a> file (extensions <code>.fasta</code>, <code>.fa</code> or <code>.fna</code>). It must contain the complete nucleotide sequence of the human genome. We will compare our <code>FASTQ</code> files against&nbsp;it;</li>
<li>A <a href="http://www.htslib.org/doc/faidx.html"><code>FASTA index</code></a> (<code>.fai</code>). It stores genomic regions as coordinates. We will use it to generate a Browser Extensible Data (<code>.bed</code>) file (see&nbsp;below);</li>
<li>A <a href="https://en.wikipedia.org/wiki/BED_(file_format)">Browser Extensible Data (<code>.bed</code>)</a> file. It stores genomic regions as coordinates, indicating the start and end of chromosomes. It is most useful when its information is chromosome-ordered and&nbsp;position-sorted;</li>
<li>Alternatively, the <code>FastQ_to_Annotation.sh</code> script can accept a <a href="https://www.ensembl.org/info/website/upload/gff.html">General Feature Format (<code>.gff</code>)</a> instead of the <code>.bed</code> file. It is used for describing genes and other features of <span class="caps">DNA</span>, <span class="caps">RNA</span> and protein&nbsp;sequences;</li>
<li>Burrows-Wheelers Aligner index files (<code>.amb</code>, <code>.ann</code>, <code>.bwt</code>, <code>.pac</code> and <code>.sa</code>). The <a href="http://bio-bwa.sourceforge.net/"><code>bwa</code> program</a> is a short read alignment tool. In other words, it identifies the location of the reads inside the <code>FASTQ</code> files. The <code>FastQ_to_Annotation.sh</code> script uses <code>bwa</code> at the second step of the pipeline. It works by efficiently using this collection of five files as a&nbsp;index.</li>
</ol>
<p><em>Why we need these files?</em> Briefly, They serve to map the genomic location of any variant we identify in our samples, as well as the genetic mutation that occurred there, which allows us to predict the possible effect(s) over the phenotype in question (in our case, MAPKi-resistance in melanoma samples). If we compare the genetic variation profile of MAPKi-susceptible samples with MAPKi-resistant samples, we could identify genetic variants associated with the resistances, and perhaps point to new directions of prognosis and new&nbsp;treatments.</p>
<p>I will now show how to obtain all these files. Remember to <strong>activate</strong> the <code>miniconda</code> that you created before if&nbsp;necessary.</p>
<p><strong><span class="caps">WARNING</span>: Several gigabytes of data will be downloaded from the internet and installed on your computer. Be sure that you have plenty or unlimited data allowances from your <span class="caps">ISP</span> and sufficient free space on your hard drive before continuing. It will take a while (several minutes to hours) until all the needed processes&nbsp;finish.</strong></p>
<h4>1. Human genome <span class="caps">FASTA</span></h4>
<div class="highlight"><pre><span></span><code><span class="c1"># First, create a subfolder into the demo folder to better organize our reference files</span>
<span class="nb">cd</span> demo
mkdir refs
<span class="nb">cd</span> refs

<span class="c1"># Download GRCh38 major release without ALT contigs and with decoy genomes (EBV and hs38d1 contig) from NCBI&#39;s FTP server</span>
curl -O ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.28_GRCh38.p13/GRCh38_major_release_seqs_for_alignment_pipelines/GCA_000001405.15_GRCh38_no_alt_plus_hs38d1_analysis_set.fna.gz
</code></pre></div>


<h4>2. Human genome <span class="caps">FASTA</span>&nbsp;index</h4>
<div class="highlight"><pre><span></span><code><span class="c1"># Download from NCBI&#39;s FTP server</span>
curl -O ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.28_GRCh38.p13/GRCh38_major_release_seqs_for_alignment_pipelines/GCA_000001405.15_GRCh38_no_alt_plus_hs38d1_analysis_set.fna.fai
</code></pre></div>


<h4>3. Human genome <span class="caps">BED</span>&nbsp;file</h4>
<div class="highlight"><pre><span></span><code><span class="c1"># Produce sorted BED file from reference genome index file obtained above</span>
awk <span class="s1">&#39;{print $1 &quot;\t0\t&quot; $2}&#39;</span> GCA_000001405.15_GRCh38_no_alt_plus_hs38d1_analysis_set.fna.fai <span class="p">|</span> sort -k1,1V -k2,2n &gt; GCA_000001405.15_GRCh38_no_alt_plus_hs38d1_analysis_set.bed
</code></pre></div>


<h4>4. Human genome <span class="caps">GFF</span> file (optional alternative to <span class="caps">BED</span>&nbsp;file)</h4>
<div class="highlight"><pre><span></span><code><span class="c1"># Download from NCBI&#39;s FTP server</span>
curl -O ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.28_GRCh38.p13/GRCh38_major_release_seqs_for_alignment_pipelines/GCA_000001405.15_GRCh38_full_analysis_set.refseq_annotation.gff.gz
</code></pre></div>


<h4>5. bwa index&nbsp;files</h4>
<div class="highlight"><pre><span></span><code>bwa index GCA_000001405.15_GRCh38_no_alt_plus_hs38d1_analysis_set.fna.gz
</code></pre></div>


<p>When you complete all of the steps in Part 1, Part 2 and in this part, your <code>demo</code> folder should have the files showed&nbsp;below</p>
<p><img alt="demo folder contents until now" src="https://antoniocampos13.github.io/images/demo_folder.PNG"></p>
<p>Now, go to the folder <a href="https://github.com/antoniocampos13/portfolio/tree/master/Unix/2020-10-01_Fastq%20to%20Annotation"><code>FastQ_to_Annotation</code> folder in my portfolio</a>, take heed of the <span class="caps">GPL</span> License and Copyright Notice, download and copy the <code>FastQ_to_Annotation.sh</code> script into your <code>demo</code> folder.</p>
<p>Thus, the only mandatory files are the <code>FastQ_to_Annotation.sh</code>, the <code>FASTQ</code> pair and the ones in the <code>refs</code> folder. If you are missing any other file, do not&nbsp;worry.</p>
<h2><span class="caps">GPL</span> License and Copyright&nbsp;Notice</h2>
<p>The <code>FastQ_to_Annotation.sh</code> script is a modified version from <a href="https://github.com/kevinblighe/ClinicalGradeDNAseq">Dr. Kevin Blighe&#8217;s original scripts</a>. Both works are licensed under <a href="https://www.gnu.org/licenses/licenses.en.html"><span class="caps">GNU</span> General Public License v3.0</a>.</p>
<h2>Conclusion of Part&nbsp;3</h2>
<p>In this part I showed how&nbsp;to:</p>
<ul>
<li>Set up <span class="caps">VEP</span> local&nbsp;cache;</li>
<li>Obtain human genome reference&nbsp;files;</li>
<li>Obtain auxiliary files needed for short read&nbsp;alignment.</li>
</ul>
<p>Finally, everything is in place for the <code>FastQ_to_Annotation.sh</code> script.</p>
<p><em><a href="https://antoniocampos13.github.io/fastq-to-annotation-part-4">Go to <span class="caps">FASTQ</span> to Annotation (Part&nbsp;4)</a></em></p>
<p><em><a href="https://antoniocampos13.github.io/fastq-to-annotation-part-1">Go back to <span class="caps">FASTQ</span> to Annotation (Part&nbsp;1)</a></em></p>
<p><em><a href="https://antoniocampos13.github.io/fastq-to-annotation-part-2">Go back to <span class="caps">FASTQ</span> to Annotation (Part&nbsp;2)</a></em></p>
<h2>References</h2>
<p><a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology&nbsp;Information</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/sra">Home - <span class="caps">SRA</span> - <span class="caps">NCBI</span></a></p>
<p><a href="https://www.ensembl.org/info/docs/tools/vep/index.html">Ensembl Variant Effect Predictor (<span class="caps">VEP</span>)</a></p>
<p><a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=BlastHelp"><span class="caps">BLAST</span> Topics | Query Input and database&nbsp;selection</a></p>
<p><a href="http://www.htslib.org/doc/faidx.html">faidx(5) manual&nbsp;page</a></p>
<p><a href="https://en.wikipedia.org/wiki/BED_(file_format)"><span class="caps">BED</span> (file format) -&nbsp;Wikipedia</a></p>
<p><a href="https://www.ensembl.org/info/website/upload/gff.html"><span class="caps">GFF</span>/<span class="caps">GTF</span> File&nbsp;Format</a></p>
<p><a href="http://bio-bwa.sourceforge.net/">Burrows-Wheeler&nbsp;Aligner</a></p>
<p><a href="https://github.com/kevinblighe/ClinicalGradeDNAseq">kevinblighe/ClinicalGradeDNAseq</a></p>
<p><a href="https://www.gnu.org/licenses/licenses.en.html"><span class="caps">GNU</span> General Public&nbsp;License</a></p>
<p><a href="http://ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.28_GRCh38.p13/GRCh38_major_release_seqs_for_alignment_pipelines/"><span class="caps">NCBI</span>&#8217;s <span class="caps">FTP</span> Server | GRCh38 Major release sequences for alignment&nbsp;pipelines</a></p>
<p><a href="http://ftp://ftp.ensembl.org/pub/release-101/fasta/homo_sapiens/dna/">Ensembl&#8217;s <span class="caps">FTP</span> Server | Homo sapiens <span class="caps">DNA</span> sequences release&nbsp;101</a></p>
  </div>
  <div class="tag-cloud">
    <p>
      <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>
      <a href="https://antoniocampos13.github.io/tag/genomic-variation.html">genomic variation</a>
      <a href="https://antoniocampos13.github.io/tag/entrez-direct.html">entrez-direct</a>
      <a href="https://antoniocampos13.github.io/tag/edirect.html">EDirect</a>
    </p>
  </div>





</article>

    <footer>
<p>
  &copy; 2020 Antonio Victor Campos Coelho - This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>
</p>
<p>
Built with <a href="http://getpelican.com" target="_blank">Pelican</a> using <a href="http://bit.ly/flex-pelican" target="_blank">Flex</a> theme
</p><p>
  <a rel="license"
     href="http://creativecommons.org/licenses/by-sa/4.0/"
     target="_blank">
    <img alt="Creative Commons License"
         title="Creative Commons License"
         style="border-width:0"
           src="https://i.creativecommons.org/l/by-sa/4.0/80x15.png"
         width="80"
         height="15"/>
  </a>
</p>    </footer>
  </main>




<script type="application/ld+json">
{
  "@context" : "http://schema.org",
  "@type" : "Blog",
  "name": " Antonio's Portfolio ",
  "url" : "https://antoniocampos13.github.io",
  "image": "https://avatars.githubusercontent.com/antoniocampos13",
  "description": "Data Science Portfolio by Antonio Victor Campos Coelho"
}
</script>


</body>
</html>