
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <meta name="HandheldFriendly" content="True" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta name="robots" content="" />

  <link href="https://fonts.googleapis.com/css2?family=Source+Code+Pro:ital,wght@0,400;0,700;1,400&family=Source+Sans+Pro:ital,wght@0,300;0,400;0,700;1,400&display=swap" rel="stylesheet">

    <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/stylesheet/style.min.css">


    <link id="pygments-light-theme" rel="stylesheet" type="text/css"
          href="https://antoniocampos13.github.io/theme/pygments/vs.min.css">


  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/fontawesome.css">
  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/brands.css">
  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/solid.css">


    <link href="https://antoniocampos13.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="Antonio's Portfolio Atom">

    <link href="https://antoniocampos13.github.io/feeds/all.rss.xml" type="application/rss+xml" rel="alternate" title="Antonio's Portfolio RSS">


  

    <!-- Chrome, Firefox OS and Opera -->
    <meta name="theme-color" content="#333333">
    <!-- Windows Phone -->
    <meta name="msapplication-navbutton-color" content="#333333">
    <!-- iOS Safari -->
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    <!-- Microsoft EDGE -->
    <meta name="msapplication-TileColor" content="#333333">

 

<meta name="author" content="Antonio Victor Campos Coelho" />
<meta name="description" content="Introduction Recently, me and my colleagues wrote a manuscript involving meta-analysis of RNA-Seq studies. One of my tasks of this project was to perform a Gene Ontology (GO) enrichment analysis: “[G]iven a set of genes that are up-regulated under certain conditions, an enrichment analysis will find which GO …" />
<meta name="keywords" content="Bioinformatics, Ensembl, BioMart, omics, data mining">


  <meta property="og:site_name" content="Antonio's Portfolio"/>
  <meta property="og:title" content="How to Query Ensembl BioMart with Python"/>
  <meta property="og:description" content="Introduction Recently, me and my colleagues wrote a manuscript involving meta-analysis of RNA-Seq studies. One of my tasks of this project was to perform a Gene Ontology (GO) enrichment analysis: “[G]iven a set of genes that are up-regulated under certain conditions, an enrichment analysis will find which GO …"/>
  <meta property="og:locale" content="en_US"/>
  <meta property="og:url" content="https://antoniocampos13.github.io/how-to-query-ensembl-biomart-with-python.html"/>
  <meta property="og:type" content="article"/>
  <meta property="article:published_time" content="2021-01-19 10:20:00-03:00"/>
  <meta property="article:modified_time" content=""/>
  <meta property="article:author" content="https://antoniocampos13.github.io/author/antonio-victor-campos-coelho.html">
  <meta property="article:section" content="Python"/>
  <meta property="article:tag" content="Bioinformatics"/>
  <meta property="article:tag" content="Ensembl"/>
  <meta property="article:tag" content="BioMart"/>
  <meta property="article:tag" content="omics"/>
  <meta property="article:tag" content="data mining"/>
  <meta property="og:image" content="https://avatars.githubusercontent.com/antoniocampos13">

  <title>Antonio's Portfolio &ndash; How to Query Ensembl BioMart with Python</title>

</head>
<body class="light-theme">
  <aside>
    <div>
      <a href="https://antoniocampos13.github.io/">
        <img src="https://avatars.githubusercontent.com/antoniocampos13" alt="Antonio's Portfolio" title="Antonio's Portfolio">
      </a>

      <h1>
        <a href="https://antoniocampos13.github.io/">Antonio's Portfolio</a>
      </h1>

<p>PhD in Genetics</p>

      <nav>
        <ul class="list">


              <li>
                <a target="_self"
                   href="https://antoniocampos13.github.io/pages/about.html#about">
                  About
                </a>
              </li>
              <li>
                <a target="_self"
                   href="https://antoniocampos13.github.io/pages/contact.html#contact">
                  Contact
                </a>
              </li>

            <li>
              <a target="_self" href="http://lattes.cnpq.br/2986394950644755" >Brazilian Lattes CV</a>
            </li>
            <li>
              <a target="_self" href="https://scholar.google.com.br/citations?user=d2ij4wUAAAAJ&hl" >Google Scholar</a>
            </li>
            <li>
              <a target="_self" href="https://orcid.org/0000-0003-2143-9701" >ORCID</a>
            </li>
            <li>
              <a target="_self" href="http://www.webofscience.com/wos/author/record/E-6795-2015" >ResearcherID Profile</a>
            </li>
        </ul>
      </nav>

      <ul class="social">
          <li>
            <a  class="sc-github" href="https://github.com/antoniocampos13/portfolio" target="_blank">
              <i class="fab fa-github"></i>
            </a>
          </li>
          <li>
            <a  class="sc-linkedin" href="https://www.linkedin.com/in/antonio-coelho-9aa338164" target="_blank">
              <i class="fab fa-linkedin"></i>
            </a>
          </li>
      </ul>
    </div>

  </aside>
  <main>

    <nav>
      <a href="https://antoniocampos13.github.io/">Home</a>

      <a href="/archives.html">Archives</a>
      <a href="/categories.html">Categories</a>
      <a href="/tags.html">Tags</a>

      <a href="https://antoniocampos13.github.io/feeds/all.atom.xml">Atom</a>

      <a href="https://antoniocampos13.github.io/feeds/all.rss.xml">RSS</a>
    </nav>

<article class="single">
  <header>
      
    <h1 id="how-to-query-ensembl-biomart-with-python">How to Query Ensembl BioMart with&nbsp;Python</h1>
    <p>
      Posted on Tue 19 January 2021 in <a href="https://antoniocampos13.github.io/category/python.html">Python</a>

    </p>
  </header>


  <div>
    <h2>Introduction</h2>
<p>Recently, me and my colleagues wrote a manuscript involving meta-analysis of <span class="caps">RNA</span>-Seq studies. One of my tasks of this project was to perform a Gene Ontology (<span class="caps">GO</span>) enrichment analysis: <a href="http://geneontology.org/docs/go-enrichment-analysis/">&#8220;[G]iven a set of genes that are up-regulated under certain conditions, an enrichment analysis will find which <span class="caps">GO</span> terms are over-represented (or under-represented) using annotations for that gene set&#8221;</a>. In other words, I could verify which cellular pathways were in action during the experimental&nbsp;conditions.</p>
<p>After I finished the <span class="caps">GO</span> analysis, I got a spreadsheet with a list of <span class="caps">GO</span> terms &mdash; brief descriptions of the cellular process performed by each pathway. When I was thinking about the pathways, I started to wonder: &#8220;which proteins participate into each pathway?&#8221; So I decided to go data mining the <a href="https://m.ensembl.org/biomart/martview">Ensembl BioMart</a> to find out those protein&nbsp;genes.</p>
<p>Ensembl is a huge project by the European Bioinformatics Institute and the Wellcome Trust Sanger Institute to provide databases of annotated genomes for several (mainly vertebrate) species. BioMart is one of their data mining tools. In this post, I will describe how I used Python to query BioMart. I will introduce a simple function to generate <a href="https://www.gnu.org/software/wget/"><code>GNU Wget</code></a> commands to retrieve query results via <a href="https://en.wikipedia.org/wiki/Representational_state_transfer">RESTful</a> access. Then, I will show how I aggregated the data to met my objective (i.e. list all genes participating in a biological pathway represented by a <span class="caps">GO</span>&nbsp;term).</p>
<p>The code and example files presented here are available in my <a href="https://github.com/antoniocampos13/portfolio/tree/master/Python/2021_01_18_How_to_Query_BioMart_Python">portfolio</a>.</p>
<h2>Prepare&nbsp;identifiers</h2>
<p>I created a project folder where I saved the Python script with this demonstration&#8217;s code (<code>ensembl_rest.py</code>) and two subfolders. The <code>data</code> folder holds the example data: a spreadsheet named <code>go_demo.xlsx</code>. The <code>src</code> folder contains the functions&#8217; code. I will write about them&nbsp;later.</p>
<div class="highlight"><pre><span></span><code>.
├── data
│   └── go_demo.xlsx
├── src
│   ├── files_to_pandas.py
│   ├── lists_ensembl.R
│   └── query_biomart.py
└── ensembl_rest.py
</code></pre></div>

<p>The spreadsheet contains just two rows of data (so that the computation can be completed quickly). They are derived from a <span class="caps">GO</span> enrichment analysis output performed by <a href="https://www.rdocumentation.org/packages/limma/versions/3.28.14/topics/goana"><code>limma</code><span class="quo">&#8216;</span>s package <code>goana</code> function</a> available for R software. The <span class="caps">GO</span> ids were the <strong>identifiers</strong> (search keywords) I used to query BioMart. You may use whatever identifier recognized by BioMart (more on that later). Just be sure they are on a nicely-named column so you can read it into&nbsp;Python.</p>
<p><img alt="Print screen of identifiers file" src="https://antoniocampos13.github.io/images/go_demo_xlsx.png"></p>
<h2>Load modules and set file&nbsp;paths</h2>
<p>These are the modules I&nbsp;used:</p>
<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">glob</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>

<span class="kn">import</span> <span class="nn">dask.dataframe</span> <span class="k">as</span> <span class="nn">dd</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">dask</span> <span class="kn">import</span> <span class="n">delayed</span>

<span class="kn">from</span> <span class="nn">src.files_to_pandas</span> <span class="kn">import</span> <span class="n">files_to_pandas</span>
<span class="kn">from</span> <span class="nn">src.query_biomart</span> <span class="kn">import</span> <span class="n">query_biomart</span>
</code></pre></div>

<p>The first four modules are from Python&#8217;s standard library. I already used <code>dask</code> <a href="https://antoniocampos13.github.io/working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-2.html">before</a>. If you do not have <code>dask</code> or <code>pandas</code> installed, do it now with <code>pip</code>. I installed <a href="https://openpyxl.readthedocs.io/en/stable/"><code>openpyxl</code></a> as well, since it serves to read/write Excel&nbsp;spreadsheets.</p>
<div class="highlight"><pre><span></span><code>pip install <span class="s2">&quot;dask[complete]&quot;</span> pandas openpyxl
</code></pre></div>

<p>Observe that I am explicitly loading the functions by prefixing the modules names with <code>src.</code> (remember that we can call every Python script a&nbsp;module).</p>
<p>Using <a href="https://docs.python.org/3/library/pathlib.html"><code>pathlib.Path</code></a> I nicely set the folders and files paths, using the project folder as the&nbsp;root.</p>
<div class="highlight"><pre><span></span><code><span class="n">project_folder</span> <span class="o">=</span> <span class="n">Path</span><span class="p">()</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span> <span class="c1"># project root</span>
<span class="n">data_folder</span> <span class="o">=</span> <span class="n">project_folder</span> <span class="o">/</span> <span class="s2">&quot;data&quot;</span>
<span class="n">go_list</span> <span class="o">=</span> <span class="n">data_folder</span> <span class="o">/</span> <span class="s2">&quot;go_demo.xlsx&quot;</span>
</code></pre></div>

<p>Then I loaded the identifiers file into a <code>pandas.DataFrame</code> with the help of <code>openpyxl</code>.</p>
<div class="highlight"><pre><span></span><code><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">go_list</span><span class="p">,</span> <span class="n">engine</span><span class="o">=</span><span class="s2">&quot;openpyxl&quot;</span><span class="p">)</span>
</code></pre></div>

<h2>Configure the queries: datasets, filters, values and&nbsp;attributes</h2>
<p>Now let&#8217;s examine an excerpt from the <code>query_biomart()</code> function code. It contains the necessary modules, the function&#8217;s arguments and some of their types hinted with the help of <code>typing</code> module:</p>
<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">re</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span>

<span class="k">def</span> <span class="nf">query_biomart</span><span class="p">(</span>
<span class="n">filter_names</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
<span class="n">values</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
<span class="n">attribute_names</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
<span class="n">dataset_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">,</span>
<span class="n">formatter</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;TSV&quot;</span><span class="p">,</span>
<span class="n">header</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="n">uniqueRows</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>

<span class="c1"># ...Function here ...</span>
</code></pre></div>

<p>There is a description of each argument&nbsp;below:</p>
<ul>
<li><code>filter_names</code> (List[str]): A list of strings that define restrictions on the&nbsp;query.</li>
<li><code>values</code> (List[str]): A list of strings containing the values that will be used for the filters. Must have same length of <code>filter_names</code>.</li>
<li><code>attribute_names</code> (List[str]): A list of strings that define the data characteristics that must be retrieved for the filtered&nbsp;data.</li>
<li><code>dataset_name</code> (str, optional): A string indicating which dataset will be queried. Each species has their respective dataset. Defaults to &#8220;hsapiens_gene_ensembl&#8221;&nbsp;(humans).</li>
<li><code>formatter</code> (str, optional): A string to indicate how output must be formatted. Options: &#8220;<code>HTML</code><span class="dquo">&#8220;</span>, &#8220;<code>CSV</code><span class="dquo">&#8220;</span>,&#8221;<code>TSV</code>&#8221; and &#8220;<code>XLS</code><span class="dquo">&#8220;</span>. Defaults to &#8220;<code>TSV</code><span class="dquo">&#8220;</span>.</li>
<li><code>header</code> (bool, optional): A Boolean indicating if the output must have a header. Defaults to <code>False</code>.</li>
<li><code>uniqueRows</code> (bool, optional): A Boolean indicating if the output must have unique rows (deduplicate data). Defaults to <code>False</code>.</li>
</ul>
<p>Thus I searched&nbsp;BioMart:</p>
<ol>
<li>Into <em>Homo sapiens</em> <strong>dataset</strong>&nbsp;(&#8220;hsapiens_gene_ensembl&#8221;),</li>
<li><strong>Filtering</strong> by <span class="caps">GO</span> terms&nbsp;(&#8220;go_parent_term&#8221;),</li>
<li>Using <span class="caps">GO</span> ids (for example, <span class="caps">GO</span>:0002790) as search keywords (<strong>values</strong>),</li>
<li>And wanted to retrieve the gene name <strong>attribute</strong>&nbsp;(&#8220;external_gene_name&#8221;).</li>
</ol>
<p>Here is an example of the function with the inputs above. All defaults were maintained, except for <code>uniqueRows</code>, which I set to <code>True</code> to remove duplicate&nbsp;data:</p>
<div class="highlight"><pre><span></span><code><span class="n">query_biomart</span><span class="p">(</span><span class="n">filter_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;go_parent_term&quot;</span><span class="p">],</span>
        <span class="n">values</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;GO:0002790&quot;</span><span class="p">],</span>
        <span class="n">attribute_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;external_gene_name&quot;</span><span class="p">],</span>
        <span class="n">uniqueRows</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>

<p><em>See the Appendix to help you see which information can be searched in&nbsp;BioMart.</em></p>
<p>The output of the function is a string &dash; a ready-to-use <code>GNU Wget</code> Unix program command. The query string inside the command has a <a href="https://m.ensembl.org/info/data/biomart/biomart_restful.html"><code>XML</code> format</a> that is sent to Ensembl&#8217;s servers. The query result will then be saved on a file named based on the values (search keywords) of the query. The extension of the file depends on the <code>formatter</code> argument.</p>
<p>I created a <code>Bash</code> script named <code>commands.sh</code> inside the <code>data</code> folder to backup and document my work. I did this with a <code>for</code> loop. In other words, Python wrote for me a series of <code>Wget</code> commands, one for each keyword in the identifiers data frame (<code>go_ids</code> column &dash; <code>df["go_ids"]</code>) alongside the desired filter and&nbsp;attribute:</p>
<div class="highlight"><pre><span></span><code><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">data_folder</span> <span class="o">/</span> <span class="s2">&quot;commands.sh&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">commands</span><span class="p">:</span>

<span class="n">commands</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;#!/bin/bash&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">go_id</span> <span class="ow">in</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;go_ids&quot;</span><span class="p">]:</span>
   <span class="n">commands</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">query_biomart</span><span class="p">(</span><span class="n">filter_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;go_parent_term&quot;</span><span class="p">],</span>
   <span class="n">values</span><span class="o">=</span><span class="p">[</span><span class="n">go_id</span><span class="p">],</span> <span class="n">attribute_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;external_gene_name&quot;</span><span class="p">],</span>
   <span class="n">uniqueRows</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div>

<p>See a print screen of the <code>commands.sh</code> file below. Notice the shebang (<code>#!</code>) line: it ensures the file can be executed by <code>Bash</code>.</p>
<p><img alt="Print screen commands.sh file" src="https://antoniocampos13.github.io/images/commands_print.png"></p>
<h2>Execute the&nbsp;queries</h2>
<p>I executed the <code>commands.sh</code> file with the help of Python&#8217;s <code>subprocess</code> library. Notice I indicated the path of the file by converting the <code>libpath.Path()</code> address to a string with <code>str()</code> and changed directories with the <code>cwd</code> argument so that the queries results would be downloaded into the <code>data</code> folder as&nbsp;well.</p>
<div class="highlight"><pre><span></span><code><span class="n">subprocess</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">data_folder</span> <span class="o">/</span> <span class="s2">&quot;commands.sh&quot;</span><span class="p">),</span> <span class="n">cwd</span><span class="o">=</span><span class="n">data_folder</span><span class="p">)</span>
</code></pre></div>

<p>Then I waited for the download completion. The print screen below shows part of the <code>wget</code> progress.</p>
<p><img alt="wget download progress" src="https://antoniocampos13.github.io/images/wget_download.png"></p>
<p>After a while, two files, named <code>GO0002790.txt</code> and <code>GO0019932.txt</code> were created, each containing a list of several genes related with the <span class="caps">GO</span>&nbsp;ids:</p>
<div class="highlight"><pre><span></span><code>.
├── data
│   ├── commands.sh
│   ├── go_demo.xlsx
│   ├── GO0002790.txt
│   ├── GO0019932.txt
│   └── lists_ensembl.xlsx
├── src
│   ├── files_to_pandas.py
│   ├── lists_ensembl.R
│   └── query_biomart.py
└── ensembl_rest.py
</code></pre></div>

<p><img alt="Result excerpt" src="https://antoniocampos13.github.io/images/go0002790.png"></p>
<h2>Label and unify data into a <code>pandas.DataFrame</code></h2>
<p>Similarly to my <a href="https://antoniocampos13.github.io/working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-2.html">previous post</a> I needed to make a relation of <span class="caps">GO</span> id &#8212;&gt; genes, so I adapted and renamed the old function I used on that occasion. The new function is named <code>files_to_pandas</code>. It takes a file and converts it into a <code>pandas.DataFrame</code> while creating a new column to accommodate the file name into a new column (in our case, the files were named after the values &dash; <span class="caps">GO</span>&nbsp;ids).</p>
<p>The idea is that I needed to create one data frame for each file. Thus, I used the <code>glob</code> method to get an iterable of file names and used <code>dask</code><span class="quo">&#8216;</span>s <a href="https://docs.dask.org/en/latest/delayed.html"><code>delayed</code></a> method to assemble the several <code>pandas.DataFrames</code> generated by the <code>files_to_pandas</code> function loop into a unified <code>dask.DataFrame</code>.</p>
<div class="highlight"><pre><span></span><code><span class="n">file_list</span> <span class="o">=</span> <span class="n">data_folder</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s2">&quot;*.txt&quot;</span><span class="p">)</span>

<span class="n">dfs</span> <span class="o">=</span> <span class="p">[</span><span class="n">delayed</span><span class="p">(</span><span class="n">files_to_pandas</span><span class="p">)(</span><span class="n">filename</span><span class="p">)</span> <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">file_list</span><span class="p">]</span>

<span class="n">ddf</span> <span class="o">=</span> <span class="n">dd</span><span class="o">.</span><span class="n">from_delayed</span><span class="p">(</span><span class="n">dfs</span><span class="p">)</span>
</code></pre></div>

<h2>Reorganize data with <code>defaultdict</code></h2>
<p>I could have ended in the previous step, but I surmised that I could condense the data frame into fewer rows, by aggregating all genes into the same row of the equivalent <span class="caps">GO</span> id. So I had the idea to use Python&#8217;s <a href="https://docs.python.org/3/library/collections.html#collections.defaultdict"><code>defaultdict</code></a>. It is a subclass of the <code>dictionary</code> class. First, I initialize an empty <code>defaultdict</code>. It will have <span class="caps">GO</span> ids as keys and a list of gene names as&nbsp;values:</p>
<div class="highlight"><pre><span></span><code><span class="n">go_to_genes</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
</code></pre></div>

<p>Then I simply had to loop through the rows of the <code>dask.DataFrame</code>, insert the keys into the <code>defaultdict</code> and appending the gene names into the value-list one by&nbsp;one.</p>
<div class="highlight"><pre><span></span><code><span class="k">for</span> <span class="n">go_id</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">ddf</span><span class="p">[</span><span class="s2">&quot;filename&quot;</span><span class="p">],</span> <span class="n">ddf</span><span class="p">[</span><span class="s2">&quot;gene_name&quot;</span><span class="p">]):</span>
    <span class="n">go_to_genes</span><span class="p">[</span><span class="n">go_id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
</code></pre></div>

<h2>Save the <code>defaultdict</code> into a new <code>pandas.DataFrame</code></h2>
<p>Finally, I created another data frame to store the contents of the <code>defaultdict</code>. Notice how I used the <code>.items()</code> method. It is the correct way to access the contents of a <code>defaultdict</code>.</p>
<div class="highlight"><pre><span></span><code><span class="n">go_genes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">go_to_genes</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;go_ids&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_name&quot;</span><span class="p">])</span>
</code></pre></div>

<h3>Error&nbsp;checking</h3>
<p>Just to be sure that all queries worked correctly, a used a simple loop to check the contents of the gene names&nbsp;column:</p>
<div class="highlight"><pre><span></span><code><span class="k">for</span> <span class="n">go_id</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">go_genes</span><span class="p">[</span><span class="s2">&quot;go_ids&quot;</span><span class="p">],</span><span class="n">go_genes</span><span class="p">[</span><span class="s2">&quot;gene_name&quot;</span><span class="p">]):</span>
    <span class="k">if</span> <span class="s2">&quot;Query&quot;</span> <span class="ow">in</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error in </span><span class="si">{</span><span class="n">go_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">continue</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Error check done.&quot;</span><span class="p">)</span>
</code></pre></div>

<p>I am using &#8220;Query&#8221; is used as an example, because BioMart errors may contain this word in their error messages, such&nbsp;as:</p>
<div class="highlight"><pre><span></span><code>&quot;Query ERROR: caught BioMart::Exception::Database: Could not connect to mysql database ...&quot;
</code></pre></div>

<p>Of course, other strings can be used. In retrospect, &#8220;<span class="caps">ERROR</span>&#8221; would have been an even better option than &#8220;Query&#8221;. If some error were detected, the loop would print the <span class="caps">GO</span> id that failed the query (due to a network error, for example), so I would have to retry the&nbsp;query.</p>
<h2>Annotate original data&nbsp;frame</h2>
<p>No errors were found, so I finally could annotate my original data frame by joining them by the <code>"go_ids"</code> column. I even created a new column (<code>"gene_string"</code>) to convert the Python lists into a nicely formatted comma-delimited string of gene&nbsp;names:</p>
<div class="highlight"><pre><span></span><code><span class="n">df_annotated</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;go_ids&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">go_genes</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;go_ids&quot;</span><span class="p">))</span>

<span class="n">df_annotated</span><span class="p">[</span><span class="s2">&quot;gene_string&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">element</span><span class="p">))</span> <span class="k">for</span> <span class="n">element</span> <span class="ow">in</span> <span class="n">df_annotated</span><span class="p">[</span><span class="s2">&quot;gene_name&quot;</span><span class="p">]]</span>
</code></pre></div>

<p>I then saved the result into a new tab (named <code>go annotated</code>) inside my <code>go_demo.xlsx</code> file:</p>
<div class="highlight"><pre><span></span><code><span class="k">with</span> <span class="n">pd</span><span class="o">.</span><span class="n">ExcelWriter</span><span class="p">(</span><span class="n">go_list</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;a&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">writer</span><span class="p">:</span>
    <span class="n">df_annotated</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s2">&quot;go annotated&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>

<h2>Conclusion</h2>
<p>In this post,&nbsp;I:</p>
<ul>
<li>Introduced a function to create customized ready-to-use query strings via <span class="caps">REST</span> <span class="caps">API</span>&nbsp;access;</li>
<li>Provided a file with the descriptors of data deposited in BioMart (<code>list_ensembl.xlsx</code>);</li>
<li>Demonstrated how to use Python to invoke <code>GNU Wget</code> to download the the queries&#8217;&nbsp;results;</li>
<li>Demonstrated how to aggregate and manipulate the queries results using <code>pandas</code>, <code>dask</code> and <code>defaultdict</code>.</li>
</ul>
<p>Feel free to use the <code>query_biomart()</code> function to data mine BioMart as you&nbsp;wish!</p>
<p><em>Subscribe to my <a href="https://antoniocampos13.github.io/feeds/all.rss.xml"><span class="caps">RSS</span> feed</a>, <a href="https://antoniocampos13.github.io/feeds/all.atom.xml">Atom feed</a> or <a href="https://t.me/joinchat/AAAAAEYrNCLK80Fh1w8nAg">Telegram channel</a> to keep you updated whenever I post new&nbsp;content.</em></p>
<h2>Appendix</h2>
<p>I prepared a spreadsheet named <code>lists_ensembl.xlsx</code> also stored into the <code>data</code> folder. Ensembl has a lot of datasets, filters and attributes, so examine the other rows if you are interested. I produced the spreadsheet with the help of <a href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/biomaRt.html"><code>biomaRt</code> R package</a>. If you prefer R over Python, be sure to try querying BioMart with it. The <code>lists_ensembl.R</code> file contains the R code that generated the&nbsp;spreadsheet.</p>
<h2>References</h2>
<p><a href="http://geneontology.org/docs/go-enrichment-analysis/"><span class="caps">GO</span> enrichment&nbsp;analysis</a></p>
<p><a href="https://m.ensembl.org/biomart/martview">Ensembl&nbsp;BioMart</a></p>
<p><a href="https://m.ensembl.org/info/data/biomart/biomart_restful.html">BioMart RESTful&nbsp;access</a></p>
<p><a href="https://www.gnu.org/software/wget/">Wget - <span class="caps">GNU</span> Project - Free Software&nbsp;Foundation</a></p>
<p><a href="https://en.wikipedia.org/wiki/Representational_state_transfer">Representational state transfer -&nbsp;Wikipedia</a></p>
<p><a href="https://www.rdocumentation.org/packages/limma/versions/3.28.14/topics/goana">goana function | R&nbsp;Documentation</a></p>
<p><a href="https://antoniocampos13.github.io/working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-2.html">Working with Cancer Genomics Cloud datasets in a PostgreSQL database (Part&nbsp;2)</a></p>
<p><a href="https://docs.python.org/3/library/pathlib.html">pathlib — Object-oriented filesystem paths &#8212; Python 3.9.1&nbsp;documentation</a></p>
<p><a href="https://openpyxl.readthedocs.io/en/stable/">openpyxl - A Python library to read/write Excel 2010 xlsx/xlsm files &mdash; openpyxl 3.0.6&nbsp;documentation</a></p>
<p><a href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/biomaRt.html">The biomaRt users&nbsp;guide</a></p>
<p><a href="https://docs.dask.org/en/latest/delayed.html">Dask  documentation -&nbsp;Delayed</a></p>
<p><a href="https://docs.python.org/3/library/collections.html#collections.defaultdict">collections — Container datatypes | Python 3.9.1&nbsp;documentation</a></p>
  </div>
  <div class="tag-cloud">
    <p>
      <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>
      <a href="https://antoniocampos13.github.io/tag/ensembl.html">Ensembl</a>
      <a href="https://antoniocampos13.github.io/tag/biomart.html">BioMart</a>
      <a href="https://antoniocampos13.github.io/tag/omics.html">omics</a>
      <a href="https://antoniocampos13.github.io/tag/data-mining.html">data mining</a>
    </p>
  </div>





</article>

    <footer>
<p>
  &copy; 2020 Antonio Victor Campos Coelho - This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/deed.en_US" target="_blank">Creative Commons Attribution-ShareAlike</a>
</p>
<p>
Built with <a href="http://getpelican.com" target="_blank">Pelican</a> using <a href="http://bit.ly/flex-pelican" target="_blank">Flex</a> theme
</p><p>
  <a rel="license"
     href="http://creativecommons.org/licenses/by-sa/4.0/"
     target="_blank">
    <img alt="Creative Commons License"
         title="Creative Commons License"
         style="border-width:0"
           src="https://i.creativecommons.org/l/by-sa/4.0/80x15.png"
         width="80"
         height="15"/>
  </a>
</p>    </footer>
  </main>




<script type="application/ld+json">
{
  "@context" : "http://schema.org",
  "@type" : "Blog",
  "name": " Antonio's Portfolio ",
  "url" : "https://antoniocampos13.github.io",
  "image": "https://avatars.githubusercontent.com/antoniocampos13",
  "description": "Data Science Portfolio by Antonio Victor Campos Coelho"
}
</script>

</body>
</html>