
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <meta name="HandheldFriendly" content="True" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta name="robots" content="" />

  <link href="https://fonts.googleapis.com/css2?family=Source+Code+Pro:ital,wght@0,400;0,700;1,400&family=Source+Sans+Pro:ital,wght@0,300;0,400;0,700;1,400&display=swap" rel="stylesheet">

    <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/stylesheet/style.min.css">


    <link id="pygments-light-theme" rel="stylesheet" type="text/css"
          href="https://antoniocampos13.github.io/theme/pygments/vs.min.css">


  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/fontawesome.css">
  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/brands.css">
  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/solid.css">


    <link href="https://antoniocampos13.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="Antonio's Portfolio Atom">

    <link href="https://antoniocampos13.github.io/feeds/all.rss.xml" type="application/rss+xml" rel="alternate" title="Antonio's Portfolio RSS">


  

    <!-- Chrome, Firefox OS and Opera -->
    <meta name="theme-color" content="#333333">
    <!-- Windows Phone -->
    <meta name="msapplication-navbutton-color" content="#333333">
    <!-- iOS Safari -->
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    <!-- Microsoft EDGE -->
    <meta name="msapplication-TileColor" content="#333333">

 

<meta name="author" content="Antonio Victor Campos Coelho" />
<meta name="description" content="Introduction ClinVar is one of the USA’s National Center for Biotechnology Information (NCBI) databases. ClinVar archives reports of relationships among human genetic variants and phenotypes (usually genetic disorders). Any organization, such as a laboratory, hospital, clinic etc can submit data to ClinVar. The core idea of ClinVar is aggregate …" />
<meta name="keywords" content="pandas, ClinVar, genomics, variants">


  <meta property="og:site_name" content="Antonio's Portfolio"/>
  <meta property="og:title" content="Parsing the ClinVar XML file with pandas"/>
  <meta property="og:description" content="Introduction ClinVar is one of the USA’s National Center for Biotechnology Information (NCBI) databases. ClinVar archives reports of relationships among human genetic variants and phenotypes (usually genetic disorders). Any organization, such as a laboratory, hospital, clinic etc can submit data to ClinVar. The core idea of ClinVar is aggregate …"/>
  <meta property="og:locale" content="en_US"/>
  <meta property="og:url" content="https://antoniocampos13.github.io/parsing-the-clinvar-xml-file-with-pandas.html"/>
  <meta property="og:type" content="article"/>
  <meta property="article:published_time" content="2023-02-04 16:25:00-03:00"/>
  <meta property="article:modified_time" content=""/>
  <meta property="article:author" content="https://antoniocampos13.github.io/author/antonio-victor-campos-coelho.html">
  <meta property="article:section" content="Python"/>
  <meta property="article:tag" content="pandas"/>
  <meta property="article:tag" content="ClinVar"/>
  <meta property="article:tag" content="genomics"/>
  <meta property="article:tag" content="variants"/>
  <meta property="og:image" content="https://avatars.githubusercontent.com/antoniocampos13">

  <title>Antonio's Portfolio &ndash; Parsing the ClinVar XML file with pandas</title>

</head>
<body class="light-theme">
  <aside>
    <div>
      <a href="https://antoniocampos13.github.io/">
        <img src="https://avatars.githubusercontent.com/antoniocampos13" alt="Antonio's Portfolio" title="Antonio's Portfolio">
      </a>

      <h1>
        <a href="https://antoniocampos13.github.io/">Antonio's Portfolio</a>
      </h1>

<p>PhD in Genetics</p>

      <nav>
        <ul class="list">


              <li>
                <a target="_self"
                   href="https://antoniocampos13.github.io/pages/about.html#about">
                  About
                </a>
              </li>
              <li>
                <a target="_self"
                   href="https://antoniocampos13.github.io/pages/contact.html#contact">
                  Contact
                </a>
              </li>

            <li>
              <a target="_self" href="http://lattes.cnpq.br/2986394950644755" >Brazilian Lattes CV</a>
            </li>
            <li>
              <a target="_self" href="https://scholar.google.com.br/citations?user=d2ij4wUAAAAJ&hl" >Google Scholar</a>
            </li>
            <li>
              <a target="_self" href="https://orcid.org/0000-0003-2143-9701" >ORCID</a>
            </li>
            <li>
              <a target="_self" href="http://www.webofscience.com/wos/author/record/E-6795-2015" >ResearcherID Profile</a>
            </li>
        </ul>
      </nav>

      <ul class="social">
          <li>
            <a  class="sc-github" href="https://github.com/antoniocampos13/portfolio" target="_blank">
              <i class="fab fa-github"></i>
            </a>
          </li>
          <li>
            <a  class="sc-linkedin" href="https://www.linkedin.com/in/antonio-coelho-9aa338164" target="_blank">
              <i class="fab fa-linkedin"></i>
            </a>
          </li>
      </ul>
    </div>

  </aside>
  <main>

    <nav>
      <a href="https://antoniocampos13.github.io/">Home</a>

      <a href="/archives.html">Archives</a>
      <a href="/categories.html">Categories</a>
      <a href="/tags.html">Tags</a>

      <a href="https://antoniocampos13.github.io/feeds/all.atom.xml">Atom</a>

      <a href="https://antoniocampos13.github.io/feeds/all.rss.xml">RSS</a>
    </nav>

<article class="single">
  <header>
      
    <h1 id="parsing-the-clinvar-xml-file-with-pandas">Parsing the ClinVar <span class="caps">XML</span> file with&nbsp;pandas</h1>
    <p>
      Posted on Sat 04 February 2023 in <a href="https://antoniocampos13.github.io/category/python.html">Python</a>

    </p>
  </header>


  <div>
    <h2>Introduction</h2>
<p><a href="https://www.ncbi.nlm.nih.gov/clinvar/intro/">ClinVar</a> is one of the <span class="caps">USA</span>&#8217;s National Center for Biotechnology Information (<span class="caps">NCBI</span>) databases. ClinVar archives reports of relationships among human genetic variants and phenotypes (usually genetic disorders). Any organization, such as a laboratory, hospital, clinic etc can submit data to ClinVar. The core idea of ClinVar is aggregate evidence for the clinical significance of any genetic variant concerning any disorder. Over 2,400 organizations contributed more than 2 million 600 thousand <a href="https://www.ncbi.nlm.nih.gov/clinvar/submitters/">records to ClinVar</a>, representing more than 1 million 600 thousand unique&nbsp;variants.</p>
<p>Anyone can freely search ClinVar through their <a href="https://www.ncbi.nlm.nih.gov/clinvar/">website</a>, using gene symbols, genomic coordinates, <a href="https://varnomen.hgvs.org/"><span class="caps">HGVS</span> expressions</a>, phenotypes, and more. If you want to perform a few queries, the online search tool does a good job. However, if you are pursuing more complex scientific questions, or are intending to download batches of data, the search tool will not suffice. Other <span class="caps">NCBI</span> databases can be queried via the command line with the <a href="https://www.ncbi.nlm.nih.gov/books/NBK179288/">Entrez Direct (EDirect) utilities</a> (in a <a href="https://antoniocampos13.github.io/fastq-to-annotation-part-2.html">previous post</a> I mention how to work with the EDirect utilities). Unfortunately, ClinVar does not currently support a batch query interface via EDirect&nbsp;utilities.</p>
<p>However, ClinVar provides <a href="https://www.ncbi.nlm.nih.gov/clinvar/docs/maintenance_use/">other approaches</a> for the access and use of their data. One of these approaches is the provisioning of the complete public data set in the form of an <a href="https://ftp.ncbi.nlm.nih.gov/pub/clinvar/xml/"><span class="caps">XML</span> file stored at the ClinVar <span class="caps">FTP</span> server</a>. The <code>ClinVarFullRelease</code> <span class="caps">XML</span> file is updated weekly, and every release happening on the first Thursday of the month is&nbsp;archived.</p>
<h2>Parsing the ClinVar <span class="caps">XML</span>&nbsp;file</h2>
<p>Recently, I started assisting my team in uploading variant/phenotype interpretations to ClinVar. I wanted to find a way to gather all our submissions into a spreadsheet so every team member could easily check whenever necessary. Thus, I downloaded the full ClinVar release <span class="caps">XML</span> file and tried to parse it with the <span class="caps">XML</span>-handling <a href="https://docs.python.org/3/library/xml.etree.elementtree.html"><code>ElementTree</code> module</a>. However, I had limited success. I could extract some information, but the output did not turn out exactly the way I was intending, so I set out to find working&nbsp;alternatives.</p>
<p>Eventually, I found out that the <a href="https://pandas.pydata.org/"><code>pandas</code> module</a> has a method to convert <span class="caps">XML</span>-stored data into traditional data frames. Moreover, since September 2022, their <code>read_xml()</code> function supports large <span class="caps">XML</span> files via the <code>iterparse</code> argument (read an excerpt of the release note <a href="https://pandas.pydata.org/docs/whatsnew/v1.5.0.html#read-xml-now-supports-large-xml-using-iterparse">here</a>).</p>
<p>The function documentation states that the <code>iterparse</code> argument is a memory-efficient method for handling big <span class="caps">XML</span> files without storing all data elements within memory. This was exactly my case, so I tried the <code>read_xml()</code> function &mdash; it worked quite&nbsp;well!</p>
<p>I wrote a small script that you can use to parse the ClinVar <span class="caps">XML</span> file. Of course, when you get acquainted with the <code>read_xml()</code>, you may use it for parsing any other <span class="caps">XML</span> you wish. I used an <a href="https://aws.amazon.com/ec2/?nc1=h_ls"><span class="caps">AWS</span> <span class="caps">EC2</span> instance</a> with 90 <span class="caps">GB</span> <span class="caps">RAM</span> while working on this tutorial. I did not try to process the ClinVar file in less powerful systems. <em>Try at your own risk</em>.</p>
<p>I uploaded the script (named <code>clinvar_pandas_xml_parser.py</code>) to <a href="https://github.com/antoniocampos13/portfolio/tree/master/Python/2023_02_04_Parsing_ClinVar_XML_with_pandas">the corresponding folder on my portfolio</a>.</p>
<h3>Downloading the full ClinVar release <span class="caps">XML</span>&nbsp;file</h3>
<p>You can download the latest <code>.gz</code>-compressed <span class="caps">XML</span> files via the following links (<strong><span class="caps">WARNING</span>:</strong> the release files are <span class="caps">HUGE</span>):</p>
<p><a href="https://ftp.ncbi.nlm.nih.gov/pub/clinvar/xml/weekly_release/ClinVarFullRelease_00-latest_weekly.xml.gz">Weekly&nbsp;release</a></p>
<p><a href="https://ftp.ncbi.nlm.nih.gov/pub/clinvar/xml/ClinVarFullRelease_00-latest.xml.gz">Monthly&nbsp;release</a></p>
<p>Go to a convenient directory on your system and download one of the files above. I downloaded the most recent monthly release and decompressed it soon&nbsp;after:</p>
<div class="highlight"><pre><span></span><code>wget https://ftp.ncbi.nlm.nih.gov/pub/clinvar/xml/ClinVarFullRelease_00-latest.xml.gz

gunzip ClinVarFullRelease_00-latest.xml.gz
</code></pre></div>

<p>If you want to check the file integrity, compare your checksum against the corresponding ClinVar-provided checksum&nbsp;file:</p>
<p><a href="https://ftp.ncbi.nlm.nih.gov/pub/clinvar/xml/weekly_release/ClinVarFullRelease_00-latest_weekly.xml.gz.md5">Weekly release (<span class="caps">MD5</span> checksum&nbsp;file)</a></p>
<p><a href="https://ftp.ncbi.nlm.nih.gov/pub/clinvar/xml/ClinVarFullRelease_00-latest.xml.gz.md5">Monthly release (<span class="caps">MD5</span> checksum&nbsp;file)</a></p>
<h3>Installing&nbsp;modules</h3>
<p>The <code>iterparse</code> argument in the <code>read_xml()</code> function was introduced in <code>pandas</code> version 1.5.0 and requires the <code>lxml</code> or <code>ElementTree</code> modules to work. In this tutorial, I will use <code>lxml</code> (the default). Therefore, install the necessary modules via <code>pip</code> or <code>conda</code> (see my <a href="https://antoniocampos13.github.io/setting-up-your-unix-computer-for-bioinformatics-analysis.html">previous post</a> on how to configure <code>conda</code> virtual environments in a Unix system). For&nbsp;example:</p>
<div class="highlight"><pre><span></span><code>conda activate env_name
conda install -c conda-forge <span class="nv">pandas</span><span class="o">=</span><span class="m">1</span>.5.0 lxml
</code></pre></div>

<h3>Running the <code>clinvar_pandas_xml_parser.py</code> script</h3>
<p>Finally, let&#8217;s walk through the script. First, I import the <code>pandas</code> module:</p>
<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
</code></pre></div>

<p>Next, I saved the <span class="caps">XML</span> file path into the <code>xml_file_path</code> object:</p>
<div class="highlight"><pre><span></span><code><span class="n">xml_file_path</span> <span class="o">=</span> <span class="s2">&quot;ClinVarFullRelease_00-latest.xml&quot;</span>
</code></pre></div>

<p>Then, I investigated the <span class="caps">XML</span> using <code>grep</code> commands to match specific strings of interest to get a feel of how the <span class="caps">XML</span> file was structured. I am sure that are better ways to assess the <span class="caps">XML</span> elements structure, but I am not an expert in <span class="caps">XML</span>&nbsp;files.</p>
<p>Through my investigation of the file, I concluded that the <code>ClinVarAssertion</code> elements within the <span class="caps">XML</span> structure contained all information I was needing at the moment. Thus, I created a Python dictionary object named <code>iterparse_dict</code> with the string &#8220;<code>ClinVarAssertion</code>&#8221; as a <em>key</em> and a Python list as its corresponding <em>value</em>:</p>
<div class="highlight"><pre><span></span><code><span class="n">iterparse_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ClinVarAssertion&quot;</span><span class="p">:</span> <span class="p">[]}</span>
</code></pre></div>

<p>The <em>key</em> represents the parent <span class="caps">XML</span> node tag. The <em>value</em> is a list containing all child or grandchild nodes, tags, or attributes at any node level inside the main <span class="caps">XML</span> node &mdash; simple as that. I chose the&nbsp;following:</p>
<div class="highlight"><pre><span></span><code><span class="n">iterparse_dict</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;ClinVarAssertion&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="s2">&quot;ID&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SubmissionName&quot;</span><span class="p">,</span>
        <span class="s2">&quot;localKey&quot;</span><span class="p">,</span>
        <span class="s2">&quot;submittedAssembly&quot;</span><span class="p">,</span>
        <span class="s2">&quot;submitter&quot;</span><span class="p">,</span>
        <span class="s2">&quot;submitterDate&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Acc&quot;</span><span class="p">,</span>
        <span class="s2">&quot;RecordStatus&quot;</span><span class="p">,</span>
        <span class="s2">&quot;OrgID&quot;</span><span class="p">,</span>
        <span class="s2">&quot;DateCreated&quot;</span><span class="p">,</span>
        <span class="s2">&quot;DateUpdated&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Version&quot;</span><span class="p">,</span>
        <span class="s2">&quot;DateLastEvaluated&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Description&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ReviewStatus&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Comment&quot;</span>
    <span class="p">]</span>
<span class="p">}</span>
</code></pre></div>

<p>Then, I passed the <code>iterparse_dict</code> as the value for the <code>iterparse</code> argument of the <code>read_xml()</code> function and stored the output as the <code>df</code> object &mdash; a <code>pandas.DataFrame</code>. The columns of the data frame will correspond to the information stored at each <code>ClinVarAssertion</code> tag,&nbsp;attributes</p>
<div class="highlight"><pre><span></span><code><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_xml</span><span class="p">(</span><span class="n">xml_file_path</span><span class="p">,</span> <span class="n">iterparse</span><span class="o">=</span><span class="n">iterparse_dict</span><span class="p">)</span>
</code></pre></div>

<p>After some time, the function returns a data frame that you can further filter to search for information. For now, I saved the data frame as a pickled&nbsp;object:</p>
<div class="highlight"><pre><span></span><code><span class="n">df</span><span class="o">.</span><span class="n">to_pickle</span><span class="p">(</span><span class="s2">&quot;pandas_parsed.pkl&quot;</span><span class="p">)</span>
</code></pre></div>

<p>At any moment, I can restore the data frame through <code>pandas</code> as well (<strong><span class="caps">REMEMBER</span>:</strong> Loading pickled data received from untrusted sources can be&nbsp;unsafe):</p>
<div class="highlight"><pre><span></span><code><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_pickle</span><span class="p">(</span><span class="s2">&quot;pandas_parsed.pkl&quot;</span><span class="p">)</span>
</code></pre></div>

<h2>Conclusion</h2>
<p>In this post, I demonstrated one way of exploring the full release of the ClinVar database, through an up-to-date <code>pandas</code> method that can deal with big <span class="caps">XML</span>&nbsp;files.</p>
<h2>Appendix</h2>
<p>A brief explanation of what each <em>value</em> in the <code>iterparse_dict</code> dictionary object&nbsp;represents:</p>
<ul>
<li><code>ID</code>: A unique numeric id representing a submission (a single submission usually contains many variant/phenotype&nbsp;interpretations).</li>
<li><code>SubmissionName</code>: A unique string representing a&nbsp;submission.</li>
<li><code>localKey</code>: The <span class="caps">HGVS</span> expression representing each variant within a single&nbsp;submission.</li>
<li><code>submittedAssembly</code>: The assembly (genome reference build) that was used for variant calling, annotation and localization. Usually is &#8220;GRCh37&#8221; or&nbsp;&#8220;GRCh38&#8221;.</li>
<li><code>submitter</code>: The organization that was responsible for the&nbsp;submission.</li>
<li><code>submitterDate</code>: The date when the submission was uploaded to&nbsp;ClinVar.</li>
<li><code>Acc</code>: A ClinVar identifier string. As stated in the <a href="https://www.ncbi.nlm.nih.gov/clinvar/docs/identifiers/">ClinVar identifiers documentation</a>: &#8220;Accession numbers in ClinVar have the pattern of 3 letters and 9 numerals. The letters are either <span class="caps">SCV</span> (think of it as Submitted record in ClinVar), <span class="caps">RCV</span> (Reference ClinVar record) or <span class="caps">VCV</span> (Variation ClinVar&nbsp;record).&#8221;</li>
<li><code>RecordStatus</code>: The status of the record, whether current, deleted or secondary&nbsp;(merged).</li>
<li><code>OrgID</code>: A unique numeric identifier for each organization that was responsible for the&nbsp;submission.</li>
<li><code>DateCreated</code>: The date when the submission was accepted and integrated into the&nbsp;database.</li>
<li><code>DateUpdated</code>: The date when the submitter updated the&nbsp;record.</li>
<li><code>Version</code>: The version assigned to a record. As stated in the <a href="https://www.ncbi.nlm.nih.gov/clinvar/docs/identifiers/">ClinVar identifiers documentation</a>: &#8220;The version number is incremented when a submitter updates a record or when the contents of a reference or variation record change because of addition to, updates of, or deletion of the <span class="caps">SCV</span> accessions on which it is&nbsp;based.&#8221;</li>
<li><code>DateLastEvaluated</code>: The date when the organization evaluated the clinical significance of any given variant in the context of any given&nbsp;phenotype.</li>
<li><code>Description</code>: A description of the clinical significance of any given variant in the context of any given phenotype, such as &#8220;Pathogenic&#8221;, &#8220;Likely pathogenic&#8221;, &#8220;Benign&#8221;,&nbsp;etc.</li>
<li><code>ReviewStatus</code>: As stated in the <a href="https://www.ncbi.nlm.nih.gov/clinvar/docs/review_status/">ClinVar review status documentation</a>: &#8220;The level of review supporting the assertion of clinical significance for the&nbsp;variation.&#8221;</li>
<li><code>Comment</code>: Any additional (free-text) comments the organization that was responsible for the submission provided regarding the interpretation of any given variant in the context of any given&nbsp;phenotype.</li>
</ul>
<p><em>Subscribe to my <a href="https://antoniocampos13.github.io/feeds/all.rss.xml"><span class="caps">RSS</span> feed</a>, <a href="https://antoniocampos13.github.io/feeds/all.atom.xml">Atom feed</a> or <a href="https://t.me/joinchat/AAAAAEYrNCLK80Fh1w8nAg">Telegram channel</a> to keep you updated whenever I post new&nbsp;content.</em></p>
<h2>References</h2>
<p><a href="https://www.ncbi.nlm.nih.gov/clinvar/intro/">ClinVar | Documentation | What is&nbsp;ClinVar?</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/clinvar/submitters/">ClinVar | Documentation |&nbsp;Submitters</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/clinvar/">ClinVar | Search&nbsp;Tool</a></p>
<p><a href="https://varnomen.hgvs.org/">Sequence Variant&nbsp;Nomenclature</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/books/NBK179288/">Entrez Direct: E-utilities on the Unix Command Line - Entrez Programming Utilities Help - <span class="caps">NCBI</span>&nbsp;Bookshelf</a></p>
<p><a href="https://antoniocampos13.github.io/fastq-to-annotation-part-2.html"><span class="caps">FASTQ</span> to Annotation (Part&nbsp;2)</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/clinvar/docs/maintenance_use/">Accessing and using data in&nbsp;ClinVar</a></p>
<p><a href="https://ftp.ncbi.nlm.nih.gov/pub/clinvar/xml/">ClinVar | <span class="caps">FTP</span> server | Index of&nbsp;/pub/clinvar/xml</a></p>
<p><a href="https://docs.python.org/3/library/xml.etree.elementtree.html">xml.etree.ElementTree — The ElementTree <span class="caps">XML</span> <span class="caps">API</span></a></p>
<p><a href="https://pandas.pydata.org/">pandas - Python Data Analysis&nbsp;Library</a></p>
<p><a href="https://pandas.pydata.org/docs/whatsnew/v1.5.0.html#read-xml-now-supports-large-xml-using-iterparse">pandas | Documentation | What’s new in 1.5.0 (September 19,&nbsp;2022)</a></p>
<p><a href="https://aws.amazon.com/ec2/?nc1=h_ls">Secure and resizable cloud compute – Amazon <span class="caps">EC2</span> – Amazon Web&nbsp;Services</a></p>
<p><a href="https://antoniocampos13.github.io/setting-up-your-unix-computer-for-bioinformatics-analysis.html">Setting Up Your Unix Computer for Bioinformatics&nbsp;Analysis</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/clinvar/docs/identifiers/">ClinVar | Documentation | Identifiers in&nbsp;ClinVar</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/clinvar/docs/review_status/">ClinVar | Documentation | Review&nbsp;status</a></p>
  </div>
  <div class="tag-cloud">
    <p>
      <a href="https://antoniocampos13.github.io/tag/pandas.html">pandas</a>
      <a href="https://antoniocampos13.github.io/tag/clinvar.html">ClinVar</a>
      <a href="https://antoniocampos13.github.io/tag/genomics.html">genomics</a>
      <a href="https://antoniocampos13.github.io/tag/variants.html">variants</a>
    </p>
  </div>





</article>

    <footer>
<p>
  &copy; 2020 Antonio Victor Campos Coelho - This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/deed.en_US" target="_blank">Creative Commons Attribution-ShareAlike</a>
</p>
<p>
Built with <a href="http://getpelican.com" target="_blank">Pelican</a> using <a href="http://bit.ly/flex-pelican" target="_blank">Flex</a> theme
</p><p>
  <a rel="license"
     href="http://creativecommons.org/licenses/by-sa/4.0/"
     target="_blank">
    <img alt="Creative Commons License"
         title="Creative Commons License"
         style="border-width:0"
           src="https://i.creativecommons.org/l/by-sa/4.0/80x15.png"
         width="80"
         height="15"/>
  </a>
</p>    </footer>
  </main>




<script type="application/ld+json">
{
  "@context" : "http://schema.org",
  "@type" : "Blog",
  "name": " Antonio's Portfolio ",
  "url" : "https://antoniocampos13.github.io",
  "image": "https://avatars.githubusercontent.com/antoniocampos13",
  "description": "Data Science Portfolio by Antonio Victor Campos Coelho"
}
</script>

</body>
</html>