
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <meta name="HandheldFriendly" content="True" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta name="robots" content="" />

  <link href="https://fonts.googleapis.com/css2?family=Source+Code+Pro:ital,wght@0,400;0,700;1,400&family=Source+Sans+Pro:ital,wght@0,300;0,400;0,700;1,400&display=swap" rel="stylesheet">

    <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/stylesheet/style.min.css">


    <link id="pygments-light-theme" rel="stylesheet" type="text/css"
          href="https://antoniocampos13.github.io/theme/pygments/vs.min.css">


  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/fontawesome.css">
  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/brands.css">
  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/solid.css">


    <link href="https://antoniocampos13.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="Antonio's Portfolio Atom">

    <link href="https://antoniocampos13.github.io/feeds/all.rss.xml" type="application/rss+xml" rel="alternate" title="Antonio's Portfolio RSS">



    <!-- Chrome, Firefox OS and Opera -->
    <meta name="theme-color" content="#333333">
    <!-- Windows Phone -->
    <meta name="msapplication-navbutton-color" content="#333333">
    <!-- iOS Safari -->
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    <!-- Microsoft EDGE -->
    <meta name="msapplication-TileColor" content="#333333">

<meta name="author" content="Antonio Victor Campos Coelho" />
<meta name="description" content="Introduction Recently I have been looking for publicly-available genomics datasets to test machine learning models in Python. During my searches for such a “toy dataset”, I came upon the Cancer Genomics Cloud (CGC) initiative. Anyone can register in CGC and have access to open access massive public datasets, like The …" />
<meta name="keywords" content="Bioinformatics, gene expression quantification, copy number variation, Windows">


<meta property="og:site_name" content="Antonio's Portfolio"/>
<meta property="og:title" content="Working with Cancer Genomics Cloud datasets in a PostgreSQL database (Part 1)"/>
<meta property="og:description" content="Introduction Recently I have been looking for publicly-available genomics datasets to test machine learning models in Python. During my searches for such a “toy dataset”, I came upon the Cancer Genomics Cloud (CGC) initiative. Anyone can register in CGC and have access to open access massive public datasets, like The …"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:url" content="https://antoniocampos13.github.io/working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-1.html"/>
<meta property="og:type" content="article"/>
<meta property="article:published_time" content="2020-10-12 12:42:00-03:00"/>
<meta property="article:modified_time" content=""/>
<meta property="article:author" content="https://antoniocampos13.github.io/author/antonio-victor-campos-coelho.html">
<meta property="article:section" content="SQL"/>
<meta property="article:tag" content="Bioinformatics"/>
<meta property="article:tag" content="gene expression quantification"/>
<meta property="article:tag" content="copy number variation"/>
<meta property="article:tag" content="Windows"/>
<meta property="og:image" content="https://avatars.githubusercontent.com/antoniocampos13">

  <title>Antonio's Portfolio &ndash; Working with Cancer Genomics Cloud datasets in a PostgreSQL database (Part 1)</title>

</head>
<body class="light-theme">
  <aside>
    <div>
      <a href="https://antoniocampos13.github.io">
        <img src="https://avatars.githubusercontent.com/antoniocampos13" alt="Antonio's Portfolio" title="Antonio's Portfolio">
      </a>

      <h1>
        <a href="https://antoniocampos13.github.io">Antonio's Portfolio</a>
      </h1>

<p>PhD in Genetics</p>

      <nav>
        <ul class="list">


              <li>
                <a target="_self"
                   href="https://antoniocampos13.github.io/pages/about.html#about">
                  About
                </a>
              </li>
              <li>
                <a target="_self"
                   href="https://antoniocampos13.github.io/pages/contact.html#contact">
                  Contact
                </a>
              </li>

            <li>
              <a target="_self" href="http://lattes.cnpq.br/2986394950644755" >Brazilian Lattes CV</a>
            </li>
            <li>
              <a target="_self" href="https://scholar.google.com.br/citations?user=d2ij4wUAAAAJ&hl" >Google Scholar</a>
            </li>
            <li>
              <a target="_self" href="https://orcid.org/0000-0003-2143-9701" >ORCID</a>
            </li>
            <li>
              <a target="_self" href="https://publons.com/researcher/2414076/antonio-victor-c-coelho" >Researcher ID/Publons</a>
            </li>
        </ul>
      </nav>

      <ul class="social">
          <li>
            <a  class="sc-github" href="https://github.com/antoniocampos13/portfolio" target="_blank">
              <i class="fab fa-github"></i>
            </a>
          </li>
          <li>
            <a  class="sc-linkedin" href="https://www.linkedin.com/in/antonio-coelho-9aa338164" target="_blank">
              <i class="fab fa-linkedin"></i>
            </a>
          </li>
      </ul>
    </div>

  </aside>
  <main>

    <nav>
      <a href="https://antoniocampos13.github.io">Home</a>

      <a href="/archives.html">Archives</a>
      <a href="/categories.html">Categories</a>
      <a href="/tags.html">Tags</a>

      <a href="https://antoniocampos13.github.io/feeds/all.atom.xml">Atom</a>

      <a href="https://antoniocampos13.github.io/feeds/all.rss.xml">RSS</a>
    </nav>

<article class="single">
  <header>
      
    <h1 id="working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-1">Working with Cancer Genomics Cloud datasets in a PostgreSQL database (Part&nbsp;1)</h1>
    <p>
      Posted on Mon 12 October 2020 in <a href="https://antoniocampos13.github.io/category/sql.html">SQL</a>

    </p>
  </header>


  <div>
    <h2>Introduction</h2>
<p>Recently I have been looking for publicly-available genomics datasets to test machine learning models in Python. During my searches for such a &#8220;toy dataset&#8221;, I came upon the <a href="http://www.cancergenomicscloud.org/">Cancer Genomics Cloud (<span class="caps">CGC</span>)</a>&nbsp;initiative.</p>
<p>Anyone can register in <span class="caps">CGC</span> and have access to open access massive public datasets, like <a href="http://cancergenome.nih.gov/">The Cancer Genomics Atlas (<span class="caps">TCGA</span>)</a>. Most individual-level genomic data can only be accessed following approval of a Data Access Request through the <a href="https://www.ncbi.nlm.nih.gov/gap/">Database of Genotypes and Phenotypes (dbGaP)</a>. For now, I guess the open data tier will suffice for this&nbsp;exercise.</p>
<p>This demonstration will be separated into two parts. Here in the first part I will provide a brief run-down of how I queried the <span class="caps">CGC</span> to obtain genomic data from cancer patients and the first steps into preparing a local PostgreSQL relational database in my&nbsp;computer.</p>
<h2>Querying the <span class="caps">CGC</span></h2>
<p>I registered at <span class="caps">CGC</span>, then I created a project in the <span class="caps">CGC</span> dashboard and went into the data browser&nbsp;tool:</p>
<p><img alt="CGC dashboard options" src="https://antoniocampos13.github.io/images/cgc_1.PNG"></p>
<p>Then, I chose the <span class="caps">TCGA</span> GRCh38 dataset and clicked on the <code>Explore selected</code> button.</p>
<p><img alt="TCGA dataset" src="https://antoniocampos13.github.io/images/cgc_2.PNG"></p>
<p>Inside the data browser, I see that there are several information&nbsp;entities:</p>
<p><img alt="TCGA information entities" src="https://antoniocampos13.github.io/images/cgc_3.PNG"></p>
<p>I clicked on the first one, <code>Cases</code> and then created a query with the following entities and&nbsp;filters:</p>
<ul>
<li><em>Entity</em><ul>
<li><em>Filters</em></li>
</ul>
</li>
<li>Case<ul>
<li>Primary site: Prostate&nbsp;Gland</li>
</ul>
</li>
<li>Diagnosis<ul>
<li>Age at&nbsp;diagnosis</li>
<li>Clinical T (<span class="caps">TNM</span>)</li>
</ul>
</li>
<li>Demographic<ul>
<li>Ethnicity</li>
<li>Race</li>
</ul>
</li>
<li>Follow up<ul>
<li>Primary therapy&nbsp;outcome</li>
</ul>
</li>
<li>File<ul>
<li>Access level:&nbsp;Open</li>
<li>Data type: Gene Level Copy Number, Gene Expression Quantification, Gene Level Copy Number&nbsp;Scores</li>
</ul>
</li>
</ul>
<p>The final query ended up like&nbsp;this:</p>
<p><img alt="Prostate cancer query" src="https://antoniocampos13.github.io/images/cgc_4.PNG"></p>
<p>In other words, the query resulted in individuals diagnosed with prostate cancer (n=237), their age at diagnosis, their demographic characteristics, their therapeutic outcomes, and their genomic data (n=1,065 files overall: 276 with raw counts of gene expression quantification, 552 with <a href="https://rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/"><span class="caps">FPKM</span></a> information, 236 from <a href="https://www.nature.com/scitable/topicpage/copy-number-variation-445/">copy number variation</a> genotyping, and a single file containing what I believe is a prostate cancer diagnosis score stratified by&nbsp;gene).</p>
<p>Then, I clicked on the <code>Copy files to project</code> and on the <code>Export</code> button and chose <code>Export as TSV</code> option. I went back to my project dashboard, clicked on the <code>Files</code> tab and downloaded&nbsp;everything.</p>
<p>I realized that the four <code>TSV</code> and the genomic data could be organized as tables on a <a href="https://en.wikipedia.org/wiki/Relational_database">relational database</a>. So I used my <a href="https://www.postgresql.org/">PostgreSQL server</a> that I have installed on computer. For this demonstration, I will use my Windows 10 <span class="caps">OS</span>, but PostgreSQL can be installed on Unix systems as well. In my portfolio I provide a Windows script and a Unix script as well containing the steps I followed to load all the data into a PostgreSQL&nbsp;database.</p>
<h2>Creating the &#8216;tcga&#8217; database into the local PostgreSQL&nbsp;server</h2>
<p>The official PostgreSQL installation instructions are <a href="https://www.postgresql.org/download/">here</a>.</p>
<p>I created a folder named <code>TCGA</code> for this project, and put the downloaded files inside a <code>data</code> subfolder. Here is a representation of my directory&nbsp;structure:</p>
<div class="highlight"><pre><span></span><code>.
└── TCGA
    ├── data
    │   ├── counts
    │   │   └── <span class="o">[</span><span class="m">200</span>+ *.counts.gz files<span class="o">]</span>
    │   ├── focal_score_by_genes
    │   ├── fpkm
    │   ├── gene_level_copy_numbers
    │   ├── cases.tsv
    │   ├── demographic.tsv
    │   ├── files.tsv
    │   └── follow_up.tsv
    ├── main_tcga.ps1
    └── main_tcga.sh
</code></pre></div>

<p>The <code>TSV</code> files are the query results and inside the folders are the files containing the genomic data (these are not their original names &#8212; I renamed them to make easier to identify the contents of each one). For now, I will use just the four <code>TSV</code> files and the <code>counts</code> folders.</p>
<p>This structure is replicated in the folder corresponding to this post in my <a href="https://github.com/antoniocampos13/portfolio/tree/master/SQL/2020_10_12_Working_Data_CGC_PostgreSQL/TCGA">portfolio</a>. The <code>main_tcga.ps1</code> and <code>main_tcga.sh</code> files contain the commands I used for this demonstration. The first is for Windows and the second for Unix&nbsp;systems.</p>
<p>Then, in the <code>TCGA</code> folder I opened a Windows PowerShell terminal and using <code>psql</code>, a terminal-based front-end to PostgreSQL, created a database named <code>tcga</code> on my local&nbsp;server:</p>
<div class="highlight"><pre><span></span><code><span class="n">psql</span> <span class="n">-U</span> <span class="n">postgres</span> <span class="n">-c</span> <span class="s2">&quot;CREATE DATABASE tcga ENCODING &#39;UTF-8&#39; LC_COLLATE &#39;English_United States&#39; LC_CTYPE &#39;English_United States&#39; TEMPLATE template0&quot;</span>
</code></pre></div>

<p>The <code>-U</code> flag serves to indicate which user will connect to the local PostgreSQL server. <code>postgres</code> is the default user created during PostgreSQL installation. The <code>-c</code> flag means that we are sending a command to the server. Note that the command is inside double quotes and strings into the command are single-quoted.In summary, this command serves to connect the <code>postgres</code> user into the server and pass a command to create the <code>tcga</code> database with certain characteristics: use <span class="caps">UTF</span>-8 codification, with English locale using the <code>template0</code> database as template, which is created by default during PostgreSQL server&nbsp;installation.</p>
<p>If during installation you provided a password to access the server, the terminal will ask for it after you press&nbsp;Enter.</p>
<h2>Creating tables in the &#8216;tcga&#8217;&nbsp;database</h2>
<p>Then, I created four tables, corresponding to each <code>TSV</code> files with the following&nbsp;command:</p>
<div class="highlight"><pre><span></span><code><span class="n">psql</span> <span class="n">-U</span> <span class="n">postgres</span> <span class="n">-d</span> <span class="n">tcga</span> <span class="n">-a</span> <span class="o">-f</span> <span class="s2">&quot;src/tcga_create_tables.sql&quot;</span>
</code></pre></div>

<p>The new friends here are <code>-d</code> and <code>-a -f</code>. <code>-d</code> is the flag that indicates the <em>database</em> I wished to connect; it is the <code>tcga</code> I created above. The <code>-a</code> serves to echo all information from the command to the terminal output so it is possible to check if the commands worked. The <code>-f</code> flag mean <em>file</em>: I am indicating that I want to pass the commands within the <code>tcga_create_tables.sql</code> file inside the <code>src</code> directory &#8212; which I created as a subfolder of the <code>TCGA</code> folder. If you are wondering how I created this file: wrote the commands in a text file and simply saved it with the <code>.sql</code> extension.</p>
<p>Below is one of the commands inside the <code>.sql</code> file:</p>
<div class="highlight"><pre><span></span><code><span class="k">CREATE</span> <span class="k">TABLE</span> <span class="n">allcases</span> <span class="p">(</span>
<span class="o">#</span><span class="n">case_id</span> <span class="nb">TEXT</span><span class="p">,</span>
<span class="o">#</span><span class="n">case_primarysite</span> <span class="nb">TEXT</span><span class="p">,</span>
<span class="o">#</span><span class="n">diagnosis</span> <span class="nb">TEXT</span><span class="p">,</span>
<span class="o">#</span><span class="n">diagnosis_ageatdiagnosis_1</span> <span class="nb">INT</span><span class="p">,</span>
<span class="o">#</span><span class="n">diagnosis_clinicalt_1</span> <span class="nb">TEXT</span>
<span class="p">);</span>
</code></pre></div>

<p>The command above creates the table <code>allcases</code> with five columns: <code>case_id</code>, <code>case_primarysite</code>, <code>diagnosis</code>, <code>diagnosis_ageatdiagnosis_1</code>, and <code>diagnosis_clinicalt_1</code>. Notice the words beside each one: they indicate the <strong>data type</strong> of the data that the column will hold. In this case I have four columns that will get text data (<code>TEXT</code>) and one that will get numbers &#8212; integers (<code>INT</code>)&nbsp;specifically.</p>
<p>Note the semicolon <code>;</code> at the end &#8212; it is a PostgreSQL requirement. It indicates the end of a command (however, if we are passing arguments through the <code>-c</code> flag the semicolon is not needed though, it is implicit within the&nbsp;flag).</p>
<p>The file have three more commands similar to the one above. The output of the second to last command should be <code>CREATE TABLE</code> messages, meaning all went well &#8212; I created tour tables inside the <code>tcga</code> database.</p>
<h2>Populating the&nbsp;tables</h2>
<p>However, they are still empty. To populate the tables, I used the four commands below, one for each table (<code>allcases</code>, <code>demographic</code>, <code>follow_up</code> and <code>allfiles</code>):</p>
<div class="highlight"><pre><span></span><code><span class="n">psql</span> <span class="n">-U</span> <span class="n">postgres</span> <span class="n">-d</span> <span class="n">tcga</span> <span class="n">-c</span> <span class="s2">&quot;\COPY allcases FROM &#39;data/cases.tsv&#39; DELIMITER E&#39;\t&#39; CSV HEADER&quot;</span>

<span class="n">psql</span> <span class="n">-U</span> <span class="n">postgres</span> <span class="n">-d</span> <span class="n">tcga</span> <span class="n">-c</span> <span class="s2">&quot;\COPY demographic FROM &#39;data/demographic.tsv&#39; DELIMITER E&#39;\t&#39; CSV HEADER&quot;</span>

<span class="n">psql</span> <span class="n">-U</span> <span class="n">postgres</span> <span class="n">-d</span> <span class="n">tcga</span> <span class="n">-c</span> <span class="s2">&quot;\COPY follow_up FROM &#39;data/follow_up.tsv&#39; DELIMITER E&#39;\t&#39; CSV HEADER&quot;</span>

<span class="n">psql</span> <span class="n">-U</span> <span class="n">postgres</span> <span class="n">-d</span> <span class="n">tcga</span> <span class="n">-c</span> <span class="s2">&quot;\COPY allfiles FROM &#39;data/files.tsv&#39; DELIMITER E&#39;\t&#39; CSV HEADER&quot;</span>
</code></pre></div>

<p>(It is good practice to separate table-creating commands of table-populating ones). In summary, the commands tell the PostgreSQL server to copy the information contained in the <code>TSV</code> files inside the <code>data</code> directory into the specified&nbsp;table.</p>
<p>The argument <code>DELIMITER E'\t'</code> means that the columns are tab-separated (delimited). This argument would be <code>DELIMITER ','</code> if the file were comma-separated or omitted&nbsp;altogether.  </p>
<p>The <code>CSV</code> indicates that we are importing a delimiter-separated file. <code>HEADER</code> means that the copied file have a header &#8212; the first line have the column titles, which <strong>must be equal</strong> to the ones specified during table creation; an error will occur otherwise. This argument must be omitted if the file does not have a&nbsp;header.</p>
<p>The output <code>COPY</code> followed by an integer (representing the number of rows copied) means that everything went well. Be careful: do not run the copy commands more than once, otherwise data duplication will&nbsp;occur.</p>
<p>With this I conclude the first part of this demonstration. In the next part I will use I will use a customized Python to help with the import of genomic data into the PostgreSQL&nbsp;database.</p>
<p><em><a href="https://antoniocampos13.github.io/working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-2.html">Go to Part 2</a>.</em></p>
<h2>Conclusion of Part&nbsp;1</h2>
<p>In this part&nbsp;I:</p>
<ul>
<li>Demonstrated how to query open access data in <span class="caps">CGC</span>;</li>
<li>Showed basic commands for importing data into tables created in a local PostgreSQL&nbsp;database.</li>
</ul>
<h2>References</h2>
<p><a href="http://www.cancergenomicscloud.org/">Cancer Genomics&nbsp;Cloud</a></p>
<p><a href="http://cancergenome.nih.gov/">The Cancer Genome Atlas&nbsp;Program</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/gap/">Home - dbGaP - <span class="caps">NCBI</span></a></p>
<p><a href="https://rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/"><span class="caps">RPKM</span>, <span class="caps">FPKM</span> and <span class="caps">TPM</span>, clearly&nbsp;explained</a></p>
<p><a href="https://www.nature.com/scitable/topicpage/copy-number-variation-445/">Copy Number Variation | Scitable by Nature&nbsp;Education</a></p>
<p><a href="https://en.wikipedia.org/wiki/Relational_database">Relational database -&nbsp;Wikipedia</a></p>
<p><a href="https://www.postgresql.org/">PostgreSQL: The world&#8217;s most advanced open source&nbsp;database</a></p>
<p><a href="https://www.postgresql.org/download/">PostgreSQL:&nbsp;Downloads</a></p>
  </div>
  <div class="tag-cloud">
    <p>
      <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>
      <a href="https://antoniocampos13.github.io/tag/gene-expression-quantification.html">gene expression quantification</a>
      <a href="https://antoniocampos13.github.io/tag/copy-number-variation.html">copy number variation</a>
      <a href="https://antoniocampos13.github.io/tag/windows.html">Windows</a>
    </p>
  </div>





</article>

    <footer>
<p>
  &copy; 2020 Antonio Victor Campos Coelho - This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>
</p>
<p>
Built with <a href="http://getpelican.com" target="_blank">Pelican</a> using <a href="http://bit.ly/flex-pelican" target="_blank">Flex</a> theme
</p><p>
  <a rel="license"
     href="http://creativecommons.org/licenses/by-sa/4.0/"
     target="_blank">
    <img alt="Creative Commons License"
         title="Creative Commons License"
         style="border-width:0"
           src="https://i.creativecommons.org/l/by-sa/4.0/80x15.png"
         width="80"
         height="15"/>
  </a>
</p>    </footer>
  </main>




<script type="application/ld+json">
{
  "@context" : "http://schema.org",
  "@type" : "Blog",
  "name": " Antonio's Portfolio ",
  "url" : "https://antoniocampos13.github.io",
  "image": "https://avatars.githubusercontent.com/antoniocampos13",
  "description": "Data Science Portfolio by Antonio Victor Campos Coelho"
}
</script>


</body>
</html>