
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <meta name="HandheldFriendly" content="True" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta name="robots" content="" />

  <link href="https://fonts.googleapis.com/css2?family=Source+Code+Pro:ital,wght@0,400;0,700;1,400&family=Source+Sans+Pro:ital,wght@0,300;0,400;0,700;1,400&display=swap" rel="stylesheet">

    <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/stylesheet/style.min.css">


    <link id="pygments-light-theme" rel="stylesheet" type="text/css"
          href="https://antoniocampos13.github.io/theme/pygments/monokai.min.css">


  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/fontawesome.css">
  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/brands.css">
  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/solid.css">


    <link href="https://antoniocampos13.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="Antonio's Portfolio Atom">

    <link href="https://antoniocampos13.github.io/feeds/all.rss.xml" type="application/rss+xml" rel="alternate" title="Antonio's Portfolio RSS">



    <!-- Chrome, Firefox OS and Opera -->
    <meta name="theme-color" content="#333333">
    <!-- Windows Phone -->
    <meta name="msapplication-navbutton-color" content="#333333">
    <!-- iOS Safari -->
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    <!-- Microsoft EDGE -->
    <meta name="msapplication-TileColor" content="#333333">

<meta name="author" content="Antonio Victor Campos Coelho" />
<meta name="description" content="Introduction In a previous post, I showed how to configure an Ubuntu system to install Bioinformatics programs. Now, using the environment I created, I will demonstrate a bash script, FasQ_to_VariantCall.sh that takes next generation sequencing (NGS) raw reads from human whole genome sequencing as input and produces variant annotation …" />
<meta name="keywords" content="Bioinformatics, genomic variation, entrez-direct, EDirect">


<meta property="og:site_name" content="Antonio's Portfolio"/>
<meta property="og:title" content="FASTQ to Variant Call (Part 3)"/>
<meta property="og:description" content="Introduction In a previous post, I showed how to configure an Ubuntu system to install Bioinformatics programs. Now, using the environment I created, I will demonstrate a bash script, FasQ_to_VariantCall.sh that takes next generation sequencing (NGS) raw reads from human whole genome sequencing as input and produces variant annotation …"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:url" content="https://antoniocampos13.github.io/drafts/fastq-to-variant-call-part-3.html"/>
<meta property="og:type" content="article"/>
<meta property="article:published_time" content="2020-10-05 18:00:00-03:00"/>
<meta property="article:modified_time" content=""/>
<meta property="article:author" content="https://antoniocampos13.github.io/author/antonio-victor-campos-coelho.html">
<meta property="article:section" content="Unix"/>
<meta property="article:tag" content="Bioinformatics"/>
<meta property="article:tag" content="genomic variation"/>
<meta property="article:tag" content="entrez-direct"/>
<meta property="article:tag" content="EDirect"/>
<meta property="og:image" content="https://avatars.githubusercontent.com/antoniocampos13">

  <title>Antonio's Portfolio &ndash; FASTQ to Variant Call (Part 3)</title>

</head>
<body class="light-theme">
  <aside>
    <div>
      <a href="https://antoniocampos13.github.io">
        <img src="https://avatars.githubusercontent.com/antoniocampos13" alt="Antonio's Portfolio" title="Antonio's Portfolio">
      </a>

      <h1>
        <a href="https://antoniocampos13.github.io">Antonio's Portfolio</a>
      </h1>

<p>PhD in Genetics</p>

      <nav>
        <ul class="list">


              <li>
                <a target="_self"
                   href="https://antoniocampos13.github.io/pages/about.html#about">
                  About
                </a>
              </li>
              <li>
                <a target="_self"
                   href="https://antoniocampos13.github.io/pages/contact.html#contact">
                  Contact
                </a>
              </li>

            <li>
              <a target="_self" href="http://lattes.cnpq.br/2986394950644755" >Brazilian Lattes CV</a>
            </li>
            <li>
              <a target="_self" href="https://orcid.org/0000-0003-2143-9701" >ORCID</a>
            </li>
            <li>
              <a target="_self" href="https://publons.com/researcher/2414076/antonio-victor-c-coelho" >Researcher ID/Publons</a>
            </li>
        </ul>
      </nav>

      <ul class="social">
          <li>
            <a  class="sc-github" href="https://github.com/antoniocampos13/portfolio" target="_blank">
              <i class="fab fa-github"></i>
            </a>
          </li>
          <li>
            <a  class="sc-linkedin" href="https://www.linkedin.com/in/antonio-coelho-9aa338164" target="_blank">
              <i class="fab fa-linkedin"></i>
            </a>
          </li>
      </ul>
    </div>

  </aside>
  <main>

    <nav>
      <a href="https://antoniocampos13.github.io">Home</a>

      <a href="/archives.html">Archives</a>
      <a href="/categories.html">Categories</a>
      <a href="/tags.html">Tags</a>

      <a href="https://antoniocampos13.github.io/feeds/all.atom.xml">Atom</a>

      <a href="https://antoniocampos13.github.io/feeds/all.rss.xml">RSS</a>
    </nav>

<article class="single">
  <header>
      
    <h1 id="fastq-to-variant-call-part-3"><span class="caps">FASTQ</span> to Variant Call (Part&nbsp;3)</h1>
    <p>
      Posted on Mon 05 October 2020 in <a href="https://antoniocampos13.github.io/category/unix.html">Unix</a>

    </p>
  </header>


  <div>
    <h2>Introduction</h2>
<p><em>In a <a href="https://antoniocampos13.github.io/setting-up-your-unix-computer-for-bioinformatics-analysis.html">previous post</a>, I showed how to configure an Ubuntu system to install Bioinformatics&nbsp;programs.</em></p>
<p><em>Now, using the environment I created, I will demonstrate a bash script, <code>FasQ_to_VariantCall.sh</code> that takes next generation sequencing (<span class="caps">NGS</span>) raw reads from human whole genome sequencing as input and produces variant annotation as output. Variant annotation is the process of identifying genetic variants in some genomic <span class="caps">DNA</span> sample, and assess, for example, if any of the found variants have any effect on phenotype, such as increased susceptibility to certain&nbsp;diseases.</em></p>
<p><em>In the first part, I showed how to search for <span class="caps">NGS</span> projects deposited in <a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (<span class="caps">NCBI</span>) databases</a> from which I can download sequencing reads later to use with the&nbsp;script.</em></p>
<p><em>In the second part, I showed how to retrieve raw genome sequencing reads in the form of <code>FASTQ</code> files, which are deposited in <a href="https://www.ncbi.nlm.nih.gov/sra"><span class="caps">SRA</span></a>.</em></p>
<p>Here in the third and final part, everything is ready for the <code>FasQ_to_VariantCall.sh</code> script demonstration using the <code>FASTQ</code> files obtained in the <a href="https://antoniocampos13.github.io/fastq-to-variant-call-part-2">previous part</a>.</p>
<h2>Final&nbsp;preparations</h2>
<h3>Installing local cache of Ensembl Variant Effect Predictor (<span class="caps">VEP</span>)</h3>
<p>The <a href="https://www.ensembl.org/info/docs/tools/vep/index.html">Ensembl Variant Effect Predictor (<span class="caps">VEP</span>)</a> is the core tool used by the script for the annotation of the effects of any variants present in the sample. It may be used online, but Ensembl recommends to download and install a local cache of all data deposited in the tool to avoid server&nbsp;overload.</p>
<p><strong>Activate</strong> the miniconda environment if needed and execute the&nbsp;command:</p>
<p><strong><span class="caps">WARNING</span></strong> about 100 <span class="caps">GB</span> of data will be downloaded from the internet and installed on your computer. Be sure that you have plenty or unlimited data allowances from your <span class="caps">ISP</span> and sufficient free space on your hard drive before&nbsp;continuing.</p>
<div class="highlight"><pre><span></span><code>vep_install -a cf -s homo_sapiens_refseq -y GRCh38 -c . –CONVERT
</code></pre></div>


<p>By default, the cache will be installed into a hidden folder in your home folder (<code>/home/&lt;YOUR_USER_NAME&gt;/.vep</code>). Thus, it is expected that different miniconda environments on the same system can find this cache. You may backup the <code>.vep/</code> folder to avoid downloading the whole thing again (but consider to download newer versions of the cache as they become available though). Remember to restore it to the equivalent location as needed (again: <code>/home/&lt;YOUR_USER_NAME&gt;/.vep</code>).</p>
<h3>Ensuring that the script will use the correct miniconda&nbsp;environment</h3>
<h2><span class="caps">GPL</span> License and Copyright&nbsp;Notice</h2>
<p>The <code>FasQ_to_VariantCall.sh</code> script is a modified version from <a href="https://github.com/kevinblighe/ClinicalGradeDNAseq">Dr. Kevin Blighe&#8217;s original scripts</a>. Both works are licensed under <a href="http://www.gnu.org/licenses"><span class="caps">GNU</span> General Public License v3.0</a>.</p>
  </div>
  <div class="tag-cloud">
    <p>
      <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>
      <a href="https://antoniocampos13.github.io/tag/genomic-variation.html">genomic variation</a>
      <a href="https://antoniocampos13.github.io/tag/entrez-direct.html">entrez-direct</a>
      <a href="https://antoniocampos13.github.io/tag/edirect.html">EDirect</a>
    </p>
  </div>





</article>

    <footer>
<p>
  &copy; 2020 Antonio Victor Campos Coelho - This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>
</p>
<p>
Built with <a href="http://getpelican.com" target="_blank">Pelican</a> using <a href="http://bit.ly/flex-pelican" target="_blank">Flex</a> theme
</p><p>
  <a rel="license"
     href="http://creativecommons.org/licenses/by-sa/4.0/"
     target="_blank">
    <img alt="Creative Commons License"
         title="Creative Commons License"
         style="border-width:0"
           src="https://i.creativecommons.org/l/by-sa/4.0/80x15.png"
         width="80"
         height="15"/>
  </a>
</p>    </footer>
  </main>




<script type="application/ld+json">
{
  "@context" : "http://schema.org",
  "@type" : "Blog",
  "name": " Antonio's Portfolio ",
  "url" : "https://antoniocampos13.github.io",
  "image": "https://avatars.githubusercontent.com/antoniocampos13",
  "description": "Data Science Portfolio by Antonio Victor Campos Coelho"
}
</script>


</body>
</html>