
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <meta name="HandheldFriendly" content="True" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta name="robots" content="" />

  <link href="https://fonts.googleapis.com/css2?family=Source+Code+Pro:ital,wght@0,400;0,700;1,400&family=Source+Sans+Pro:ital,wght@0,300;0,400;0,700;1,400&display=swap" rel="stylesheet">

    <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/stylesheet/style.min.css">


    <link id="pygments-light-theme" rel="stylesheet" type="text/css"
          href="https://antoniocampos13.github.io/theme/pygments/vs.min.css">


  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/fontawesome.css">
  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/brands.css">
  <link rel="stylesheet" type="text/css" href="https://antoniocampos13.github.io/theme/font-awesome/css/solid.css">


    <link href="https://antoniocampos13.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="Antonio's Portfolio Atom">

    <link href="https://antoniocampos13.github.io/feeds/all.rss.xml" type="application/rss+xml" rel="alternate" title="Antonio's Portfolio RSS">


  

    <!-- Chrome, Firefox OS and Opera -->
    <meta name="theme-color" content="#333333">
    <!-- Windows Phone -->
    <meta name="msapplication-navbutton-color" content="#333333">
    <!-- iOS Safari -->
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    <!-- Microsoft EDGE -->
    <meta name="msapplication-TileColor" content="#333333">

 

    <meta name="author" content="Antonio Victor Campos Coelho" />
    <meta name="description" content="Data Science Portfolio by Antonio Victor Campos Coelho" />
  <meta property="og:site_name" content="Antonio's Portfolio"/>
  <meta property="og:type" content="blog"/>
  <meta property="og:title" content="Antonio's Portfolio"/>
  <meta property="og:description" content="Data Science Portfolio by Antonio Victor Campos Coelho"/>
  <meta property="og:locale" content="en_US"/>
  <meta property="og:url" content="https://antoniocampos13.github.io"/>
  <meta property="og:image" content="https://avatars.githubusercontent.com/antoniocampos13">

  <title>Antonio's Portfolio &ndash; Posts by Antonio Victor Campos Coelho:</title>

</head>
<body class="light-theme">
  <aside>
    <div>
      <a href="https://antoniocampos13.github.io/">
        <img src="https://avatars.githubusercontent.com/antoniocampos13" alt="Antonio's Portfolio" title="Antonio's Portfolio">
      </a>

      <h1>
        <a href="https://antoniocampos13.github.io/">Antonio's Portfolio</a>
      </h1>

<p>PhD in Genetics</p>

      <nav>
        <ul class="list">


              <li>
                <a target="_self"
                   href="https://antoniocampos13.github.io/pages/about.html#about">
                  About
                </a>
              </li>
              <li>
                <a target="_self"
                   href="https://antoniocampos13.github.io/pages/contact.html#contact">
                  Contact
                </a>
              </li>

            <li>
              <a target="_self" href="http://lattes.cnpq.br/2986394950644755" >Brazilian Lattes CV</a>
            </li>
            <li>
              <a target="_self" href="https://scholar.google.com.br/citations?user=d2ij4wUAAAAJ&hl" >Google Scholar</a>
            </li>
            <li>
              <a target="_self" href="https://orcid.org/0000-0003-2143-9701" >ORCID</a>
            </li>
            <li>
              <a target="_self" href="http://www.webofscience.com/wos/author/record/E-6795-2015" >ResearcherID Profile</a>
            </li>
        </ul>
      </nav>

      <ul class="social">
          <li>
            <a  class="sc-github" href="https://github.com/antoniocampos13/portfolio" target="_blank">
              <i class="fab fa-github"></i>
            </a>
          </li>
          <li>
            <a  class="sc-linkedin" href="https://www.linkedin.com/in/antonio-coelho-9aa338164" target="_blank">
              <i class="fab fa-linkedin"></i>
            </a>
          </li>
      </ul>
    </div>

  </aside>
  <main>

    <nav>
      <a href="https://antoniocampos13.github.io/">Home</a>

      <a href="/archives.html">Archives</a>
      <a href="/categories.html">Categories</a>
      <a href="/tags.html">Tags</a>

      <a href="https://antoniocampos13.github.io/feeds/all.atom.xml">Atom</a>

      <a href="https://antoniocampos13.github.io/feeds/all.rss.xml">RSS</a>
    </nav>



<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/parallelization-with-r.html#parallelization-with-r">Parallelization with&nbsp;R</a></h2>
    <p>
      Posted on Mon 31 July 2023 in <a href="https://antoniocampos13.github.io/category/r.html">R</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>,              <a href="https://antoniocampos13.github.io/tag/gene-expression.html">gene expression</a>,              <a href="https://antoniocampos13.github.io/tag/edger.html">edgeR</a>,              <a href="https://antoniocampos13.github.io/tag/furrr.html">furrr</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p>Sometimes, some computations can be carried out in parallel. Certain large tasks can be divided into independent ones, allowing them to be solved at the same time, rather than waiting for each task to be solved&nbsp;sequentially.</p>
<p>I find the native R parallel functions such as <code>mclapply()</code>, or those …</p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/parallelization-with-r.html#parallelization-with-r">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/genomic-plots-with-circlize.html#genomic-plots-with-circlize">Genomic plots with&nbsp;circlize</a></h2>
    <p>
      Posted on Sat 29 April 2023 in <a href="https://antoniocampos13.github.io/category/r.html">R</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/circlize.html">circlize</a>,              <a href="https://antoniocampos13.github.io/tag/genomics.html">genomics</a>,              <a href="https://antoniocampos13.github.io/tag/data-visualization.html">data visualization</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p>Genomics is undoubtedly a complex science. The human genome is huge, with more than 3 billion base pairs, about 20,000 protein-coding genes, several millions of variants, and many more interesting characteristics. The visualization of genomic/omics data is challenging due to the sheer volume of information. Circular plots …</p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/genomic-plots-with-circlize.html#genomic-plots-with-circlize">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/parsing-the-clinvar-xml-file-with-pandas.html#parsing-the-clinvar-xml-file-with-pandas">Parsing the ClinVar <span class="caps">XML</span> file with&nbsp;pandas</a></h2>
    <p>
      Posted on Sat 04 February 2023 in <a href="https://antoniocampos13.github.io/category/python.html">Python</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/pandas.html">pandas</a>,              <a href="https://antoniocampos13.github.io/tag/clinvar.html">ClinVar</a>,              <a href="https://antoniocampos13.github.io/tag/genomics.html">genomics</a>,              <a href="https://antoniocampos13.github.io/tag/variants.html">variants</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p><a href="https://www.ncbi.nlm.nih.gov/clinvar/intro/">ClinVar</a> is one of the <span class="caps">USA</span>&#8217;s National Center for Biotechnology Information (<span class="caps">NCBI</span>) databases. ClinVar archives reports of relationships among human genetic variants and phenotypes (usually genetic disorders). Any organization, such as a laboratory, hospital, clinic etc can submit data to ClinVar. The core idea of ClinVar is aggregate …</p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/parsing-the-clinvar-xml-file-with-pandas.html#parsing-the-clinvar-xml-file-with-pandas">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/opening-files-of-size-larger-than-ram-with-pandas.html#opening-files-of-size-larger-than-ram-with-pandas">Opening files of size larger than <span class="caps">RAM</span> with&nbsp;pandas</a></h2>
    <p>
      Posted on Mon 27 June 2022 in <a href="https://antoniocampos13.github.io/category/python.html">Python</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/pandas.html">pandas</a>,              <a href="https://antoniocampos13.github.io/tag/genomics.html">Genomics</a>,              <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p>Dealing with big files is a routine for everyone working in genomics. <span class="caps">FASTQ</span>, <span class="caps">VCF</span>, <span class="caps">BAM</span>, and <span class="caps">GTF</span>/<span class="caps">GFF3</span> files, to name a few, can range from some hundreds of megabytes to several gigabytes in size. Usually, we can use cloud services to configure computing instances with a lot of …</p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/opening-files-of-size-larger-than-ram-with-pandas.html#opening-files-of-size-larger-than-ram-with-pandas">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/integrating-r-and-python-with-reticulate.html#integrating-r-and-python-with-reticulate">Integrating R and Python with&nbsp;reticulate</a></h2>
    <p>
      Posted on Sun 20 March 2022 in <a href="https://antoniocampos13.github.io/category/r.html">R</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/reticulate.html">reticulate</a>,              <a href="https://antoniocampos13.github.io/tag/gff.html">gff</a>,              <a href="https://antoniocampos13.github.io/tag/genomicranges.html">GenomicRanges</a>,              <a href="https://antoniocampos13.github.io/tag/pyranges.html">pyranges</a>,              <a href="https://antoniocampos13.github.io/tag/bsgenome.html">BSgenome</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p><a href="https://rstudio.github.io/reticulate/">reticulate</a> is an R package that allows interoperability between R and Python. I recently discovered this package, and I have been excited to efficiently run Python scripts inside an R session, bringing the best of both&nbsp;worlds.</p>
<p>In this post, I will demonstrate <code>reticulate</code> with two scripts. First, I …</p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/integrating-r-and-python-with-reticulate.html#integrating-r-and-python-with-reticulate">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/genomic-analysis-with-hail.html#genomic-analysis-with-hail">Genomic Analysis With&nbsp;Hail</a></h2>
    <p>
      Posted on Fri 09 July 2021 in <a href="https://antoniocampos13.github.io/category/python.html">Python</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>,              <a href="https://antoniocampos13.github.io/tag/genomics.html">Genomics</a>,              <a href="https://antoniocampos13.github.io/tag/hail.html">Hail</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p>Hello, long time no see! Since I lasted posted, many things happened. Since March I have been working as Post-Doc Researcher, hired by the <a href="https://www.einstein.br/Pages/Home.aspx">Hospital Israelita Albert Einstein (<span class="caps">HIAE</span>, São Paulo, Brazil)</a> to work for the Projeto Genomas Raros (&#8220;Rare Genomes Project&#8221;, <span class="caps">GRAR</span> from here on), a public-private partnership …</p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/genomic-analysis-with-hail.html#genomic-analysis-with-hail">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/making-an-interactive-map-with-shiny-and-leaflet-in-r.html#making-an-interactive-map-with-shiny-and-leaflet-in-r">Making an Interactive Map with Shiny and Leaflet in&nbsp;R</a></h2>
    <p>
      Posted on Thu 18 February 2021 in <a href="https://antoniocampos13.github.io/category/r.html">R</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/shiny.html">shiny</a>,              <a href="https://antoniocampos13.github.io/tag/leaflet.html">leaflet</a>,              <a href="https://antoniocampos13.github.io/tag/data-visualization.html">data visualization</a>,              <a href="https://antoniocampos13.github.io/tag/web-app.html">web app</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p><a href="https://shiny.rstudio.com/">Shiny</a> is a R package developed and maintained by the <a href="https://rstudio.com/">RStudio</a> team. With Shiny, anyone can build interactive web apps to help data visualization. Here I present a simple template of an interactive Brazilian map displaying fictitious allelic frequencies with samples sizes across the country. It is a useful …</p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/making-an-interactive-map-with-shiny-and-leaflet-in-r.html#making-an-interactive-map-with-shiny-and-leaflet-in-r">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/how-to-query-ensembl-biomart-with-python.html#how-to-query-ensembl-biomart-with-python">How to Query Ensembl BioMart with&nbsp;Python</a></h2>
    <p>
      Posted on Tue 19 January 2021 in <a href="https://antoniocampos13.github.io/category/python.html">Python</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>,              <a href="https://antoniocampos13.github.io/tag/ensembl.html">Ensembl</a>,              <a href="https://antoniocampos13.github.io/tag/biomart.html">BioMart</a>,              <a href="https://antoniocampos13.github.io/tag/omics.html">omics</a>,              <a href="https://antoniocampos13.github.io/tag/data-mining.html">data mining</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p>Recently, me and my colleagues wrote a manuscript involving meta-analysis of <span class="caps">RNA</span>-Seq studies. One of my tasks of this project was to perform a Gene Ontology (<span class="caps">GO</span>) enrichment analysis: <a href="http://geneontology.org/docs/go-enrichment-analysis/">&#8220;[G]iven a set of genes that are up-regulated under certain conditions, an enrichment analysis will find which <span class="caps">GO …</span></a></p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/how-to-query-ensembl-biomart-with-python.html#how-to-query-ensembl-biomart-with-python">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/machine-learning-with-python-supervised-classification-of-tcga-prostate-cancer-data-part-1-making-features-datasets.html#machine-learning-with-python-supervised-classification-of-tcga-prostate-cancer-data-part-1-making-features-datasets">Machine Learning with Python: Supervised Classification of <span class="caps">TCGA</span> Prostate Cancer Data (Part 1 - Making Features&nbsp;Datasets)</a></h2>
    <p>
      Posted on Thu 05 November 2020 in <a href="https://antoniocampos13.github.io/category/python.html">Python</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>,              <a href="https://antoniocampos13.github.io/tag/gene-expression.html">gene expression</a>,              <a href="https://antoniocampos13.github.io/tag/machine-learning.html">machine learning</a>,              <a href="https://antoniocampos13.github.io/tag/supervised-classification.html">supervised classification</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p>In a <a href="https://antoniocampos13.github.io/working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-1.html">previous post</a>, I showed how to retrieve <a href="https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga">The Cancer Genome Atlas (<span class="caps">TCGA</span>)</a> data from the <a href="http://www.cancergenomicscloud.org/">Cancer Genomics Cloud (<span class="caps">CGC</span>) platform</a>. I downloaded gene expression quantification data, created a relational database with PostgreSQL, and created a dataset uniting the raw quantification data for 675 differentially expressed genes <a href="https://antoniocampos13.github.io/differential-expression-analysis-with-edger-in-r.html#differential-expression-analysis-with-edger-in-r">identified …</a></p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/machine-learning-with-python-supervised-classification-of-tcga-prostate-cancer-data-part-1-making-features-datasets.html#machine-learning-with-python-supervised-classification-of-tcga-prostate-cancer-data-part-1-making-features-datasets">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/machine-learning-with-python-supervised-classification-of-tcga-prostate-cancer-data-part-2-making-a-model.html#machine-learning-with-python-supervised-classification-of-tcga-prostate-cancer-data-part-2-making-a-model">Machine Learning with Python: Supervised Classification of <span class="caps">TCGA</span> Prostate Cancer Data (Part 2 - Making a&nbsp;Model)</a></h2>
    <p>
      Posted on Thu 05 November 2020 in <a href="https://antoniocampos13.github.io/category/python.html">Python</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>,              <a href="https://antoniocampos13.github.io/tag/gene-expression.html">gene expression</a>,              <a href="https://antoniocampos13.github.io/tag/machine-learning.html">machine learning</a>,              <a href="https://antoniocampos13.github.io/tag/supervised-classification.html">supervised classification</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p><em>In a <a href="https://antoniocampos13.github.io/working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-1.html">previous post</a>, I showed how to retrieve <a href="https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga">The Cancer Genome Atlas (<span class="caps">TCGA</span>)</a> data from the <a href="http://www.cancergenomicscloud.org/">Cancer Genomics Cloud (<span class="caps">CGC</span>) platform</a>. I downloaded gene expression quantification data, created a relational database with PostgreSQL, and created a dataset uniting the raw quantification data for 675 differentially expressed genes <a href="https://antoniocampos13.github.io/differential-expression-analysis-with-edger-in-r.html#differential-expression-analysis-with-edger-in-r">identified …</a></em></p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/machine-learning-with-python-supervised-classification-of-tcga-prostate-cancer-data-part-2-making-a-model.html#machine-learning-with-python-supervised-classification-of-tcga-prostate-cancer-data-part-2-making-a-model">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/differential-expression-analysis-with-edger-in-r.html#differential-expression-analysis-with-edger-in-r">Differential Expression Analysis with edgeR in&nbsp;R</a></h2>
    <p>
      Posted on Mon 26 October 2020 in <a href="https://antoniocampos13.github.io/category/r.html">R</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>,              <a href="https://antoniocampos13.github.io/tag/gene-expression.html">gene expression</a>,              <a href="https://antoniocampos13.github.io/tag/edger.html">edgeR</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p>In my <a href="https://antoniocampos13.github.io/data-manipulation-with-r.html#data-manipulation-with-r">previous post</a> I demonstrated how to organize the <a href="https://antoniocampos13.github.io/working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-1.html"><span class="caps">CGC</span> prostate cancer data</a> to a format suited to differential expression analysis (<span class="caps">DEA</span>).</p>
<p>Nowadays, <span class="caps">DEA</span> usually arises from high-throughput sequencing of a collection (library) of <span class="caps">RNA</span> molecules expressed by single cells or tissue given their conditions upon collection and …</p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/differential-expression-analysis-with-edger-in-r.html#differential-expression-analysis-with-edger-in-r">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/data-manipulation-with-r.html#data-manipulation-with-r">Data manipulation with&nbsp;R</a></h2>
    <p>
      Posted on Mon 19 October 2020 in <a href="https://antoniocampos13.github.io/category/r.html">R</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>,              <a href="https://antoniocampos13.github.io/tag/gene-expression.html">gene expression</a>,              <a href="https://antoniocampos13.github.io/tag/sql.html">SQL</a>,              <a href="https://antoniocampos13.github.io/tag/postgresql.html">PostgreSQL</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p>In my <a href="https://antoniocampos13.github.io/working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-1.html#working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-1">previous post</a> I demonstrated how to obtain a prostate cancer dataset with genomic information in the form of gene expression quantification and created a local PostgreSQL database to hold the&nbsp;data.</p>
<p>Here, I will use R to connect to the PostgreSQL database, retrieve and then prepare the …</p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/data-manipulation-with-r.html#data-manipulation-with-r">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/meta-analysis-and-meta-regression-with-r.html#meta-analysis-and-meta-regression-with-r">Meta-analysis and Meta-regression with&nbsp;R</a></h2>
    <p>
      Posted on Tue 13 October 2020 in <a href="https://antoniocampos13.github.io/category/r.html">R</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/meta-analysis.html">meta-analysis</a>,              <a href="https://antoniocampos13.github.io/tag/statistical-analysis.html">statistical analysis</a>,              <a href="https://antoniocampos13.github.io/tag/covid-19.html">COVID-19</a>,              <a href="https://antoniocampos13.github.io/tag/sars-cov-2.html">SARS-CoV-2</a>,              <a href="https://antoniocampos13.github.io/tag/acute-kidney-injury.html">acute kidney injury</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p>On December 2019, reports from severe acute respiratory syndrome in Wuhan, China, were linked to a novel coronavirus, now known as <span class="caps">SARS</span>-CoV-2, and the disease it causes was termed coronavirus disease 2019 (<span class="caps">COVID</span>-19).</p>
<p>The World Health Organization declared the <span class="caps">COVID</span>-19 outbreak a Public Health Emergency of …</p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/meta-analysis-and-meta-regression-with-r.html#meta-analysis-and-meta-regression-with-r">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-1.html#working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-1">Working with Cancer Genomics Cloud datasets in a PostgreSQL database (Part&nbsp;1)</a></h2>
    <p>
      Posted on Mon 12 October 2020 in <a href="https://antoniocampos13.github.io/category/sql.html">SQL</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>,              <a href="https://antoniocampos13.github.io/tag/gene-expression-quantification.html">gene expression quantification</a>,              <a href="https://antoniocampos13.github.io/tag/copy-number-variation.html">copy number variation</a>,              <a href="https://antoniocampos13.github.io/tag/windows.html">Windows</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p>Recently I have been looking for publicly-available genomics datasets to test machine learning models in Python. During my searches for such a &#8220;toy dataset&#8221;, I came upon the <a href="http://www.cancergenomicscloud.org/">Cancer Genomics Cloud (<span class="caps">CGC</span>)</a>&nbsp;initiative.</p>
<p>Anyone can register in <span class="caps">CGC</span> and have access to open access massive public datasets, like <a href="http://cancergenome.nih.gov/">The …</a></p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-1.html#working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-1">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-2.html#working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-2">Working with Cancer Genomics Cloud datasets in a PostgreSQL database (Part&nbsp;2)</a></h2>
    <p>
      Posted on Mon 12 October 2020 in <a href="https://antoniocampos13.github.io/category/sql.html">SQL</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>,              <a href="https://antoniocampos13.github.io/tag/gene-expression-quantification.html">gene expression quantification</a>,              <a href="https://antoniocampos13.github.io/tag/copy-number-variation.html">copy number variation</a>,              <a href="https://antoniocampos13.github.io/tag/windows.html">Windows</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p><em>Recently I have been looking for publicly-available genomics datasets to test machine learning models in Python. During my searches for such a &#8220;toy dataset&#8221;, I came upon the <a href="http://www.cancergenomicscloud.org/">Cancer Genomics Cloud (<span class="caps">CGC</span>)</a>&nbsp;initiative.</em></p>
<p><em>Anyone can register in <span class="caps">CGC</span> and have access to open access massive public datasets, like <a href="http://cancergenome.nih.gov/">The …</a></em></p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-2.html#working-with-cancer-genomics-cloud-datasets-in-a-postgresql-database-part-2">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/fastq-to-annotation-part-4.html#fastq-to-annotation-part-4"><span class="caps">FASTQ</span> to Annotation (Part&nbsp;4)</a></h2>
    <p>
      Posted on Tue 06 October 2020 in <a href="https://antoniocampos13.github.io/category/unix.html">Unix</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>,              <a href="https://antoniocampos13.github.io/tag/genomic-variation.html">genomic variation</a>,              <a href="https://antoniocampos13.github.io/tag/entrez-direct.html">entrez-direct</a>,              <a href="https://antoniocampos13.github.io/tag/edirect.html">EDirect</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p><em>In a <a href="https://antoniocampos13.github.io/setting-up-your-unix-computer-for-bioinformatics-analysis.html">previous post</a>, I showed how to configure an Ubuntu system to install Bioinformatics&nbsp;programs.</em></p>
<p><em>Now, using the environment I created, I will demonstrate a bash script, <code>FastQ_to_Annotation.sh</code> that takes next generation sequencing (<span class="caps">NGS</span>) raw reads from human whole genome sequencing as input and produces …</em></p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/fastq-to-annotation-part-4.html#fastq-to-annotation-part-4">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/fastq-to-annotation-part-3.html#fastq-to-annotation-part-3"><span class="caps">FASTQ</span> to Annotation (Part&nbsp;3)</a></h2>
    <p>
      Posted on Mon 05 October 2020 in <a href="https://antoniocampos13.github.io/category/unix.html">Unix</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>,              <a href="https://antoniocampos13.github.io/tag/genomic-variation.html">genomic variation</a>,              <a href="https://antoniocampos13.github.io/tag/entrez-direct.html">entrez-direct</a>,              <a href="https://antoniocampos13.github.io/tag/edirect.html">EDirect</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p><em>In a <a href="https://antoniocampos13.github.io/setting-up-your-unix-computer-for-bioinformatics-analysis.html">previous post</a>, I showed how to configure an Ubuntu system to install Bioinformatics&nbsp;programs.</em></p>
<p><em>Now, using the environment I created, I will demonstrate a bash script, <code>FastQ_to_Annotation.sh</code> that takes next generation sequencing (<span class="caps">NGS</span>) raw reads from human whole genome sequencing as input and produces …</em></p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/fastq-to-annotation-part-3.html#fastq-to-annotation-part-3">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/fastq-to-annotation-part-2.html#fastq-to-annotation-part-2"><span class="caps">FASTQ</span> to Annotation (Part&nbsp;2)</a></h2>
    <p>
      Posted on Fri 02 October 2020 in <a href="https://antoniocampos13.github.io/category/unix.html">Unix</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>,              <a href="https://antoniocampos13.github.io/tag/genomic-variation.html">genomic variation</a>,              <a href="https://antoniocampos13.github.io/tag/entrez-direct.html">entrez-direct</a>,              <a href="https://antoniocampos13.github.io/tag/edirect.html">EDirect</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p><em>In a <a href="https://antoniocampos13.github.io/setting-up-your-unix-computer-for-bioinformatics-analysis">previous post</a>, I showed how to configure an Ubuntu system to install Bioinformatics&nbsp;programs.</em></p>
<p><em>Now, using the environment I created, I will demonstrate a bash script, <code>FastQ_to_Annotation.sh</code> that takes next generation sequencing (<span class="caps">NGS</span>) raw reads from human whole genome sequencing as input and produces …</em></p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/fastq-to-annotation-part-2.html#fastq-to-annotation-part-2">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/fastq-to-annotation-part-1.html#fastq-to-annotation-part-1"><span class="caps">FASTQ</span> to Annotation (Part&nbsp;1)</a></h2>
    <p>
      Posted on Thu 01 October 2020 in <a href="https://antoniocampos13.github.io/category/unix.html">Unix</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>,              <a href="https://antoniocampos13.github.io/tag/genomic-variation.html">genomic variation</a>,              <a href="https://antoniocampos13.github.io/tag/entrez-direct.html">entrez-direct</a>,              <a href="https://antoniocampos13.github.io/tag/edirect.html">EDirect</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p>In my <a href="https://antoniocampos13.github.io/setting-up-your-unix-computer-for-bioinformatics-analysis.html">previous post</a>, I showed how to configure an Ubuntu system to install Bioinformatics&nbsp;programs.</p>
<p>Now, using the environment I created, I will demonstrate a bash script, <code>FastQ_to_Annotation.sh</code> that takes next generation sequencing (<span class="caps">NGS</span>) raw reads from human whole genome sequencing as input and produces …</p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/fastq-to-annotation-part-1.html#fastq-to-annotation-part-1">
          Continue reading
        </a>
  </div>
  <hr />
</article>
<article>
  <header>
    <h2><a href="https://antoniocampos13.github.io/setting-up-your-unix-computer-for-bioinformatics-analysis.html#setting-up-your-unix-computer-for-bioinformatics-analysis">Setting Up Your Unix Computer for Bioinformatics&nbsp;Analysis</a></h2>
    <p>
      Posted on Wed 30 September 2020 in <a href="https://antoniocampos13.github.io/category/unix.html">Unix</a>

          &#8226; Tagged with
              <a href="https://antoniocampos13.github.io/tag/bioinformatics.html">Bioinformatics</a>
    </p>
  </header>
  <div>
      <div><h2>Introduction</h2>
<p>In this post I will show how I set up my Unix machine to use Bioinformatics programs and tools. I am currently using Ubuntu 20.04 <span class="caps">LTS</span> (Focal Fossa) 64-bit on a <a href="https://www.digitalocean.com/community/posts/trying-the-new-wsl-2-its-fast-windows-subsystem-for-linux">Windows Subsystem for Linux (<span class="caps">WSL2</span>)</a> on Windows 10, so no <span class="caps">GUI</span>&nbsp;today!</p>
<p>The code and files used …</p></div>
        <br>
        <a class="btn"
           href="https://antoniocampos13.github.io/setting-up-your-unix-computer-for-bioinformatics-analysis.html#setting-up-your-unix-computer-for-bioinformatics-analysis">
          Continue reading
        </a>
  </div>
</article>

  <div class="pagination">
  </div>



    <footer>
<p>
  &copy; 2020 Antonio Victor Campos Coelho - This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/deed.en_US" target="_blank">Creative Commons Attribution-ShareAlike</a>
</p>
<p>
Built with <a href="http://getpelican.com" target="_blank">Pelican</a> using <a href="http://bit.ly/flex-pelican" target="_blank">Flex</a> theme
</p><p>
  <a rel="license"
     href="http://creativecommons.org/licenses/by-sa/4.0/"
     target="_blank">
    <img alt="Creative Commons License"
         title="Creative Commons License"
         style="border-width:0"
           src="https://i.creativecommons.org/l/by-sa/4.0/80x15.png"
         width="80"
         height="15"/>
  </a>
</p>    </footer>
  </main>




<script type="application/ld+json">
{
  "@context" : "http://schema.org",
  "@type" : "Blog",
  "name": " Antonio's Portfolio ",
  "url" : "https://antoniocampos13.github.io",
  "image": "https://avatars.githubusercontent.com/antoniocampos13",
  "description": "Data Science Portfolio by Antonio Victor Campos Coelho"
}
</script>

</body>
</html>